// @frontegg/rest-api/node/subscriptions/managedSubscriptions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ManagedSubscriptionApi = void 0;
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
class ManagedSubscriptionApi extends _BaseApiClient.BaseApiClient {
  constructor(...args) {
    super(...args);
    this.getManagedSubscription = async subscriptionId => {
      return this.get(`${_constants.urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}`);
    };
    this.getManagedSubscriptions = async () => {
      return this.get(_constants.urls.subscriptions.managedSubscriptions.v1);
    };
    this.cancelManagedSubscription = async subscriptionId => {
      return this.put(`${_constants.urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}/cancellations/`, {});
    };
    this.renewManagedSubscription = async subscriptionId => {
      return this.put(`${_constants.urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}/renewals/`, {});
    };
    this.updateManagedSubscription = async (subscriptionId, {
      paymentMethodId,
      planId
    }) => {
      return this.put(`${_constants.urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}`, {
        paymentMethodId,
        planId
      });
    };
  }
}
exports.ManagedSubscriptionApi = ManagedSubscriptionApi;
