// @frontegg/rest-api/node/subscriptions/subscriptions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SubscriptionApi = void 0;
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
class SubscriptionApi extends _BaseApiClient.BaseApiClient {
  constructor(...args) {
    super(...args);
    this.getSubscriptions = async () => {
      return this.get(_constants.urls.subscriptions.billing.subscriptions.v1);
    };
    this.getSubscription = async subscriptionId => {
      return this.get(`${_constants.urls.subscriptions.billing.subscriptions.v1}/${subscriptionId}`);
    };
    this.cancelSubscription = async subscriptionId => {
      return this.put(`${_constants.urls.subscriptions.billing.subscriptions.v1}/${subscriptionId}/cancellations/`, {});
    };
    this.renewSubscription = async subscriptionId => {
      return this.put(`${_constants.urls.subscriptions.billing.subscriptions.v1}/${subscriptionId}/renewals/`, {});
    };
    this.updateSubscription = async (subscriptionId, {
      paymentMethodId,
      planId
    }) => {
      return this.put(`${_constants.urls.subscriptions.billing.subscriptions.v1}/${subscriptionId}`, {
        paymentMethodId,
        planId
      });
    };
  }
}
exports.SubscriptionApi = SubscriptionApi;
