// @frontegg/rest-api/node/directory/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DirectoryApi = void 0;
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
class DirectoryApi extends _BaseApiClient.BaseApiClient {
  constructor(appName) {
    super(appName);
    this.getConfigs = async () => {
      return this.get(`${_constants.urls.directory.v1}`);
    };
    this.updateConfiguration = async (id, body) => {
      return this.patch(`${_constants.urls.directory.v1}/${id}`, body);
    };
    this.createConfiguration = async body => {
      return this.post(`${_constants.urls.directory.v1}`, body);
    };
    this.deleteConfiguration = async (id, deleteAll) => {
      return this.delete(`${_constants.urls.directory.v1}/${id}`, {
        deleteAll
      });
    };
    this.countConfigurationUsers = async id => {
      return this.get(`${_constants.urls.directory.temp}/${id}/Users/count`);
    };
  }

  /**
   * Returns all SCIM2 configurations
   *
   * ``authorized user``
   */
}
exports.DirectoryApi = DirectoryApi;
var _default = new DirectoryApi('default');
exports.default = _default;
