// @frontegg/redux-store/node/auth/MSP/dialogs/state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initialState = exports.default = void 0;
var _proxy = require("../../../toolkit/proxy");
var _restApi = require("@frontegg/rest-api");
const initialState = {
  createSubAccountDialog: {
    loading: false,
    open: false,
    error: false,
    jwt: '',
    canSelectParentAccount: false
  },
  updateSubAccountDialog: {
    accountId: '',
    accountName: '',
    loading: false,
    open: false,
    error: false,
    jwt: ''
  },
  disableSubAccountManagementDialog: {
    accountId: '',
    accountName: '',
    loading: false,
    open: false,
    error: false
  },
  enableSubAccountManagementDialog: {
    accountId: '',
    accountName: '',
    loading: false,
    open: false,
    error: false
  },
  updateSubAccountHierarchySettingsDialog: {
    accountId: '',
    accountName: '',
    subAccountAccessType: _restApi.SubAccountAccessTypeEnum.DEFAULT_OFF,
    loading: false,
    open: false,
    error: false
  },
  deleteSubAccountDialog: {
    accountId: '',
    accountName: '',
    isParentAccount: false,
    loading: false,
    open: false,
    error: false,
    jwt: ''
  },
  addUsersToAccountDialog: {
    loading: false,
    open: false,
    accountName: '',
    accountId: '',
    permissionToGiveAccessToSubAccounts: false,
    jwt: ''
  },
  deleteUserFromAccountDialog: {
    userId: '',
    userName: '',
    accountId: '',
    loading: false,
    open: false,
    jwt: ''
  },
  editUserRolesDialog: {
    loading: false,
    open: false,
    jwt: '',
    accountId: '',
    userId: '',
    userName: '',
    userRoles: []
  }
};
exports.initialState = initialState;
var _default = overrideState => (0, _proxy.createProxy)(initialState, overrideState);
exports.default = _default;
