// @frontegg/js/node/FronteggApp/FronteggApp.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FronteggApp = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var FronteggReduxStore = _interopRequireWildcard(require("@frontegg/redux-store"));
var FronteggTypes = _interopRequireWildcard(require("@frontegg/types"));
var _utils = require("../utils");
var _AppHolder = require("../AppHolder");
var FronteggRestApi = _interopRequireWildcard(require("@frontegg/rest-api"));
var _version = _interopRequireDefault(require("../version"));
var _mockFlagsList = require("../utils/mockFlagsList");
var _utils2 = require("./utils");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var FronteggApp = /*#__PURE__*/function () {
  function FronteggApp(_options, name) {
    var _this = this,
      _this$options$authOpt3,
      _options$authOptions,
      _options$authOptions2;
    var iframeRendering = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
    var allowMultipleRenderers = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
    (0, _classCallCheck2["default"])(this, FronteggApp);
    this.name = void 0;
    this.iframeRendering = void 0;
    this.options = void 0;
    this.loading = true;
    this.cdnUrl = "https://assets.frontegg.com/admin-box/".concat(_version["default"].cdnVersion);
    this.assetsUrl = "https://adminBox.frontegg.com/admin-box/".concat(_version["default"].cdnVersion);
    this.customElementName = '';
    this.adminPortalEl = void 0;
    this.adminPortalRenderer = void 0;
    this.loginBoxEl = void 0;
    this.checkoutDialogEl = void 0;
    this.checkoutDialogRenderer = void 0;
    this.adminPortalContainer = void 0;
    this.loginBoxContainer = void 0;
    this.checkoutDialogContainer = void 0;
    this.loginBoxRenderer = void 0;
    this.store = void 0;
    this.loadingListeners = [];
    this.fetchClient = void 0;
    this.apiClient = void 0;
    this.updateLocalizations = void 0;
    this.loadAdminBoxMetadata = /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
      var _data$rows$0$configur, _data$rows, _data$rows$, data, metadata;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            if (_this.options.metadata) {
              _context.next = 21;
              break;
            }
            _context.prev = 1;
            _context.next = 4;
            return (0, FronteggReduxStore.retryIfNeeded)(function () {
              return _this.fetchClient.get('/metadata', {
                entityName: 'adminBox'
              });
            }, FronteggReduxStore.DEFAULT_RETRY_CONFIG);
          case 4:
            data = _context.sent;
            metadata = (_data$rows$0$configur = data == null ? void 0 : (_data$rows = data.rows) == null ? void 0 : (_data$rows$ = _data$rows[0]) == null ? void 0 : _data$rows$.configuration) != null ? _data$rows$0$configur : {};
            _context.prev = 6;
            _context.next = 9;
            return (0, _utils2.getOverriddenMetadata)(metadata, _this.fetchClient, FronteggReduxStore.DEFAULT_RETRY_CONFIG);
          case 9:
            _this.options.metadata = _context.sent;
            _context.next = 15;
            break;
          case 12:
            _context.prev = 12;
            _context.t0 = _context["catch"](6);
            console.error('failed to get admin portal overridden metadata', _context.t0);
          case 15:
            _context.next = 21;
            break;
          case 17:
            _context.prev = 17;
            _context.t1 = _context["catch"](1);
            console.error('failed to get admin portal metadata', _context.t1);
            _this.dispatchErrorByRequestName(_context.t1, FronteggReduxStore.REQUEST_NAME.LOAD_ADMIN_BOX_METADATA);
          case 21:
            FronteggTypes.Metadata.set(_this.options.metadata, _this.name);
          case 22:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[1, 17], [6, 12]]);
    }));
    this.requestAuthorize = /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
      var _this$options$authOpt, _this$options$authOpt2;
      var shouldRequestHostedLoginAuthorize;
      return _regenerator["default"].wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            shouldRequestHostedLoginAuthorize = (_this$options$authOpt = _this.options.authOptions) == null ? void 0 : (_this$options$authOpt2 = _this$options$authOpt.hostedLoginOptions) == null ? void 0 : _this$options$authOpt2.loadUserOnFirstLoad;
            if (!_this.options.builderMode && _this.options.framework !== 'nextjs') {
              if (_this.options.hostedLoginBox && !_this.options.customLoginBox) {
                if (shouldRequestHostedLoginAuthorize) {
                  _this.store.dispatch({
                    type: 'auth/requestHostedLoginAuthorizeV2',
                    payload: {
                      shouldRedirectToLogin: false,
                      firstTime: true
                    }
                  });
                }
              } else {
                _this.store.dispatch({
                  type: 'auth/requestAuthorize',
                  payload: true
                });
              }
            }
          case 2:
          case "end":
            return _context2.stop();
        }
      }, _callee2);
    }));
    this.loadFeatureFlags = /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
      var flags;
      return _regenerator["default"].wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            if (!(_this.options.previewMode || _this.options.builderMode)) {
              _context3.next = 4;
              break;
            }
            FronteggRestApi.FeatureFlags.set(_mockFlagsList.mockFlagsList, _this.name);
            _context3.next = 16;
            break;
          case 4:
            _context3.prev = 4;
            _context3.next = 7;
            return (0, FronteggReduxStore.retryIfNeeded)(function () {
              return _this.apiClient.featureFlags.loadFeatureFlags();
            }, FronteggReduxStore.DEFAULT_RETRY_CONFIG);
          case 7:
            flags = _context3.sent;
            if (_this.options.overrideFeatureFlags) {
              Object.assign(flags, _this.options.overrideFeatureFlags);
            }
            FronteggRestApi.FeatureFlags.set(flags, _this.name);
            _context3.next = 16;
            break;
          case 12:
            _context3.prev = 12;
            _context3.t0 = _context3["catch"](4);
            console.error('failed to get admin portal feature flags', _context3.t0);
            _this.dispatchErrorByRequestName(_context3.t0, FronteggReduxStore.REQUEST_NAME.LOAD_FEATURE_FLAGS);
          case 16:
          case "end":
            return _context3.stop();
        }
      }, _callee3, null, [[4, 12]]);
    }));
    this.setFeatureFlagsForPreview = function (previewFeatureFlags) {
      FronteggRestApi.FeatureFlags.set(Object.assign({}, _mockFlagsList.mockFlagsList, previewFeatureFlags), _this.name);
    };
    this.queryFeatureFlags = function (flags) {
      return FronteggRestApi.FeatureFlags.getFeatureFlags(flags, _this.name);
    };
    this.updateLocalizationsSetter = function (localizationUpdateFn) {
      _this.updateLocalizations = localizationUpdateFn;
    };
    this.getEntitlementsFromStore = function () {
      var _this$store$getState$;
      return (_this$store$getState$ = _this.store.getState().auth.user) == null ? void 0 : _this$store$getState$.entitlements;
    };
    this.getUserFromStore = function () {
      return _this.store.getState().auth.user;
    };
    this.getEntitlementsQueryData = function (customAttributes) {
      var user = _this.getUserFromStore();
      var entitlements = _this.getEntitlementsFromStore();
      var attributes = {
        custom: customAttributes,
        jwt: user
      };
      return {
        entitlements: entitlements,
        attributes: attributes
      };
    };
    this.renderUsersTable = /*#__PURE__*/function () {
      var _ref4 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(container, props, options) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return _this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context4.sent;
              return _context4.abrupt("return", FronteggAdminPortal.renderUsersTable(container, {
                injector: _this,
                options: _this.options,
                container: container,
                props: props,
                themeOptions: options.themeOptions
              }));
            case 4:
            case "end":
              return _context4.stop();
          }
        }, _callee4);
      }));
      return function (_x, _x2, _x3) {
        return _ref4.apply(this, arguments);
      };
    }();
    this.renderInviteUserDialog = /*#__PURE__*/function () {
      var _ref5 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(container, props, options) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) switch (_context5.prev = _context5.next) {
            case 0:
              _context5.next = 2;
              return _this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context5.sent;
              return _context5.abrupt("return", FronteggAdminPortal.renderInviteUserDialog(container, {
                injector: _this,
                options: _this.options,
                container: container,
                props: props,
                themeOptions: options.themeOptions
              }));
            case 4:
            case "end":
              return _context5.stop();
          }
        }, _callee5);
      }));
      return function (_x4, _x5, _x6) {
        return _ref5.apply(this, arguments);
      };
    }();
    this.renderChangePasswordForm = /*#__PURE__*/function () {
      var _ref6 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(container, props, options) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) switch (_context6.prev = _context6.next) {
            case 0:
              _context6.next = 2;
              return _this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context6.sent;
              return _context6.abrupt("return", FronteggAdminPortal.renderChangePasswordForm(container, {
                injector: _this,
                options: _this.options,
                props: props,
                themeOptions: options.themeOptions
              }));
            case 4:
            case "end":
              return _context6.stop();
          }
        }, _callee6);
      }));
      return function (_x7, _x8, _x9) {
        return _ref6.apply(this, arguments);
      };
    }();
    this.renderProfilePage = /*#__PURE__*/function () {
      var _ref7 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(container, props, options) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) switch (_context7.prev = _context7.next) {
            case 0:
              _context7.next = 2;
              return _this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context7.sent;
              return _context7.abrupt("return", FronteggAdminPortal.renderProfilePage(container, {
                injector: _this,
                options: _this.options,
                props: props,
                themeOptions: options.themeOptions,
                localizations: options.localizations
              }));
            case 4:
            case "end":
              return _context7.stop();
          }
        }, _callee7);
      }));
      return function (_x10, _x11, _x12) {
        return _ref7.apply(this, arguments);
      };
    }();
    this.renderEditEmailForm = /*#__PURE__*/function () {
      var _ref8 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(container, props, options) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) switch (_context8.prev = _context8.next) {
            case 0:
              _context8.next = 2;
              return _this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context8.sent;
              return _context8.abrupt("return", FronteggAdminPortal.renderEditEmailForm(container, {
                injector: _this,
                options: _this.options,
                props: props,
                themeOptions: options.themeOptions,
                localizations: options.localizations
              }));
            case 4:
            case "end":
              return _context8.stop();
          }
        }, _callee8);
      }));
      return function (_x13, _x14, _x15) {
        return _ref8.apply(this, arguments);
      };
    }();
    this.getUsersTableStore = /*#__PURE__*/function () {
      var _ref9 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(options) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) switch (_context9.prev = _context9.next) {
            case 0:
              _context9.next = 2;
              return _this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context9.sent;
              return _context9.abrupt("return", FronteggAdminPortal.useUsersTableStore(options));
            case 4:
            case "end":
              return _context9.stop();
          }
        }, _callee9);
      }));
      return function (_x16) {
        return _ref9.apply(this, arguments);
      };
    }();
    this.getInviteUserDialogStore = /*#__PURE__*/function () {
      var _ref10 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(options) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) switch (_context10.prev = _context10.next) {
            case 0:
              _context10.next = 2;
              return _this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context10.sent;
              return _context10.abrupt("return", FronteggAdminPortal.useInviteUserDialogStore(options));
            case 4:
            case "end":
              return _context10.stop();
          }
        }, _callee10);
      }));
      return function (_x17) {
        return _ref10.apply(this, arguments);
      };
    }();
    var appName = (0, _utils.formatName)(name);
    var customElementName = "frontegg-app-".concat(appName);
    this.iframeRendering = iframeRendering;
    this.name = appName;
    this.options = Object.assign({}, _options, {
      contextOptions: Object.assign({
        requestCredentials: 'include'
      }, _options.contextOptions)
    });
    this.fetchClient = new FronteggRestApi.FetchClient(this.name);
    this.apiClient = (0, FronteggRestApi.createApiClient)(this.name);
    if (this.options.cdnUrl) {
      this.cdnUrl = this.options.cdnUrl;
    }
    if (this.options.assetsUrl) {
      this.assetsUrl = this.options.assetsUrl;
    }
    FronteggRestApi.ContextHolder["for"](this.name).setCdnUrl(this.cdnUrl);
    FronteggRestApi.ContextHolder["for"](this.name).setAssetsUrl(this.assetsUrl);
    this.customElementName = customElementName;
    var authOptions = this.options.authOptions ? Object.assign({}, this.options.authOptions, {
      hostedLoginBox: this.options.hostedLoginBox
    }) : {
      hostedLoginBox: this.options.hostedLoginBox
    };
    authOptions.onRedirectTo = _options.onRedirectTo;
    authOptions.disableSilentRefresh = (_this$options$authOpt3 = this.options.authOptions) == null ? void 0 : _this$options$authOpt3.disableSilentRefresh;
    if (this.options.store) {
      this.store = this.options.store;
      if (!allowMultipleRenderers && this.options.store.getState().root.appName !== name) {
        throw Error('Mismatch in store names');
      }
    } else {
      var _this$options$authOpt4, _this$options$auditsO;
      this.store = (0, FronteggReduxStore.createStore)({
        context: this.options.contextOptions,
        name: this.name,
        builderMode: this.options.builderMode,
        previewMode: this.options.previewMode,
        storeHolder: this,
        urlStrategy: this.options.urlStrategy,
        initialState: {
          auth: Object.assign({}, authOptions, (_this$options$authOpt4 = this.options.authOptions) != null ? _this$options$authOpt4 : {}),
          auditLogs: Object.assign({}, (_this$options$auditsO = this.options.auditsOptions) != null ? _this$options$auditsO : {})
        }
      });
    }
    _AppHolder.AppHolder.setInstance(appName, this);
    var contextHolder = FronteggRestApi.ContextHolder["for"](appName);
    contextHolder.setAppName(appName);
    this.setEntitlementsOptions(_options.entitlementsOptions);
    this.setSessionContext({
      enableSessionPerTenant: (_options$authOptions = _options.authOptions) == null ? void 0 : _options$authOptions.enableSessionPerTenant,
      skipUserLoading: (_options$authOptions2 = _options.authOptions) == null ? void 0 : _options$authOptions2.skipUserLoading
    });

    // pay attention: basename is used from the router base name, then if it's empty it will be the frontegg options basename
    var basename = _options.basename;
    if (basename) {
      contextHolder.setBasename(basename);
    }
  }

  /**
   * Set entitlements related options in context holder according to frontegg options
   * @param entitlementsOptions - entitlements options from frontegg options
   */
  (0, _createClass2["default"])(FronteggApp, [{
    key: "setEntitlementsOptions",
    value: function setEntitlementsOptions(entitlementsOptions) {
      FronteggRestApi.ContextHolder["for"](this.name).setEntitlementsOptions(Object.assign({}, entitlementsOptions));
    }
  }, {
    key: "setSessionContext",
    value: function setSessionContext(sessionContext) {
      FronteggRestApi.ContextHolder["for"](this.name).setSessionContext(Object.assign({}, sessionContext));
    }
  }, {
    key: "dispatchErrorByRequestName",
    value: function dispatchErrorByRequestName(error, requestName) {
      if ((error == null ? void 0 : error.statusCode) < 500) return;
      this.store.dispatch({
        type: 'auth/setErrorByRequestName',
        payload: {
          requestName: requestName,
          traceId: (0, FronteggReduxStore.errorTraceId)(error)
        }
      });
    }
  }, {
    key: "initContainers",
    value: function () {
      var _initContainers = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(elements) {
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) switch (_context11.prev = _context11.next) {
            case 0:
              this.adminPortalEl = elements.adminPortalEl;
              this.loginBoxEl = elements.loginBoxEl;
              this.checkoutDialogEl = elements.checkoutDialogEl;
              this.adminPortalContainer = elements.adminPortalContainer;
              this.loginBoxContainer = elements.loginBoxContainer;
              this.checkoutDialogContainer = elements.checkoutDialogContainer;
              try {
                if (this.options.framework !== 'nextjs') {
                  document.body.classList.add('frontegg-loading');
                }
              } catch (e) {
                console.error('Failed to add frontegg-loading to body classList');
              }
              _context11.next = 9;
              return Promise.all([this.loadFeatureFlags(), this.loadAdminBoxMetadata(), this.requestAuthorize()]);
            case 9:
              if (!this.options.previewMode && !this.options.customLoginBox) {
                this.loadLoginBox();
                (0, _utils2.loadGTM)(this.name);
              }
              if (!this.options.lazyLoadAdminPortal) {
                this.loadScript('FronteggAdminPortal');
              }
            case 11:
            case "end":
              return _context11.stop();
          }
        }, _callee11, this);
      }));
      function initContainers(_x18) {
        return _initContainers.apply(this, arguments);
      }
      return initContainers;
    }()
  }, {
    key: "updateMetadata",
    value: function updateMetadata(metadata) {
      FronteggTypes.Metadata.set(metadata, this.name);
      if (metadata != null && metadata.localizations) {
        var _this$updateLocalizat;
        (_this$updateLocalizat = this.updateLocalizations) == null ? void 0 : _this$updateLocalizat.call(this, metadata.localizations);
      }
    }
  }, {
    key: "addOnLoadedListener",
    value: function addOnLoadedListener(listener) {
      if (!this.loading) {
        listener();
        return;
      }
      this.loadingListeners.push(listener);
    }
  }, {
    key: "ready",
    value: function ready(listener) {
      if (this.loading) {
        listener();
      } else {
        this.addOnLoadedListener(listener);
      }
    }
  }, {
    key: "loginWithRedirect",
    value: function loginWithRedirect() {
      var additionalParams = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      this.store.dispatch({
        type: 'auth/setState',
        payload: {
          isLoading: true
        }
      });
      this.store.dispatch({
        type: 'auth/requestHostedLoginAuthorize',
        payload: additionalParams
      });
    }
  }, {
    key: "loginWithRedirectV2",
    value: function loginWithRedirectV2(payload) {
      this.store.dispatch({
        type: 'auth/requestHostedLoginAuthorizeV2',
        payload: payload
      });
    }
  }, {
    key: "logout",
    value: function logout(callback) {
      this.store.dispatch({
        type: 'auth/setState',
        payload: {
          isLoading: true
        }
      });
      this.store.dispatch({
        type: 'auth/logout',
        payload: callback
      });
    }
  }, {
    key: "loadScript",
    value: function loadScript(component) {
      var _this2 = this;
      return new Promise(function (resolve, reject) {
        // TODO: prevent this behavior if development
        var scriptHolder = window;
        if (scriptHolder[component]) {
          resolve(scriptHolder[component]);
          return;
        }

        

        var entries = {
          FronteggLoginBox: 'login-box/index.js',
          FronteggAdminPortal: 'admin-portal/index.js'
        };
        var script = document.createElement('script');
        script.src = "".concat(_this2.cdnUrl, "/").concat(entries[component]);
        script.setAttribute('id', "".concat(component, "Script"));
        script.setAttribute('cdn-url', _this2.cdnUrl);
        script.onload = function () {
          if (scriptHolder["".concat(component, "Init")]) {
            resolve(scriptHolder["".concat(component, "Init")]({
              FronteggRestApi: FronteggRestApi,
              FronteggTypes: FronteggTypes,
              FronteggReduxStore: FronteggReduxStore
            }));
          } else {
            resolve(scriptHolder[component]);
          }
        };
        script.onerror = function (e) {
          // eslint-disable-next-line no-console
          console.log('Profile page error', e);
          reject(e);
        };
        document.body.append(script);
      });
    }
  }, {
    key: "loadLoginBox",
    value: function () {
      var _loadLoginBox = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12() {
        var FronteggLoginBox;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) switch (_context12.prev = _context12.next) {
            case 0:
              _context12.next = 2;
              return this.loadScript('FronteggLoginBox');
            case 2:
              FronteggLoginBox = _context12.sent;
              FronteggLoginBox.render(this.loginBoxEl, {
                injector: this,
                options: this.options
              });
              this.loading = false;
              this.loadingListeners.forEach(function (listener) {
                try {
                  listener();
                } catch (e) {
                  console.error('Failed to call loadLoginBoxListener', e);
                }
              });
              this.loadingListeners = [];
            case 7:
            case "end":
              return _context12.stop();
          }
        }, _callee12, this);
      }));
      function loadLoginBox() {
        return _loadLoginBox.apply(this, arguments);
      }
      return loadLoginBox;
    }()
  }, {
    key: "showAdminPortal",
    value: function () {
      var _showAdminPortal = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13(options) {
        var _ref11, isMultiApp, _this$queryFeatureFla, _this$queryFeatureFla2, multiAppsFlag, FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee13$(_context13) {
          while (1) switch (_context13.prev = _context13.next) {
            case 0:
              _ref11 = options != null ? options : {
                isMultiApp: false
              }, isMultiApp = _ref11.isMultiApp;
              _this$queryFeatureFla = this.queryFeatureFlags(['multi-apps-admin-portal']), _this$queryFeatureFla2 = (0, _slicedToArray2["default"])(_this$queryFeatureFla, 1), multiAppsFlag = _this$queryFeatureFla2[0];
              FronteggRestApi.ContextHolder["for"](this.name).setRequestSource(FronteggRestApi.RequestSource.AdminPortal);
              _context13.next = 5;
              return this.loadScript('FronteggAdminPortal');
            case 5:
              FronteggAdminPortal = _context13.sent;
              this.adminPortalRenderer = FronteggAdminPortal.render(this.adminPortalEl, {
                injector: this,
                options: Object.assign({
                  isMultiApp: multiAppsFlag && isMultiApp
                }, this.options)
              });
            case 7:
            case "end":
              return _context13.stop();
          }
        }, _callee13, this);
      }));
      function showAdminPortal(_x19) {
        return _showAdminPortal.apply(this, arguments);
      }
      return showAdminPortal;
    }()
    /**
     * Open the hosted portal must be called synchronously with user interaction
     *
     * @param newTab - open the portal in a new tab = true
     */
  }, {
    key: "openHostedAdminPortal",
    value: function openHostedAdminPortal() {
      var newTab = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
      var baseUrl = this.options.contextOptions.baseUrl;
      var hostedPortalUrl = "".concat(baseUrl, "/oauth/portal");
      if (newTab) {
        window.open(hostedPortalUrl, '_blank');
        return;
      }
      window.location.href = hostedPortalUrl;
    }
  }, {
    key: "hideAdminPortal",
    value: function hideAdminPortal() {
      try {
        var _this$adminPortalRend;
        var searchParams = new URLSearchParams(window.location.search);
        (_this$adminPortalRend = this.adminPortalRenderer) == null ? void 0 : _this$adminPortalRend.unmount();
        // unmount is removing url search params, therefore we need to restore them after unmount
        (0, _utils.restoreSearchParams)(searchParams);
        FronteggRestApi.ContextHolder["for"](this.name).setRequestSource(null);
      } catch (e) {
        console.error('Failed to unmount admin portal', e);
      }
    }
  }, {
    key: "showCheckoutDialog",
    value: function () {
      var _showCheckoutDialog = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee14(opts) {
        var FronteggAdminPortal;
        return _regenerator["default"].wrap(function _callee14$(_context14) {
          while (1) switch (_context14.prev = _context14.next) {
            case 0:
              _context14.next = 2;
              return this.loadScript('FronteggAdminPortal');
            case 2:
              FronteggAdminPortal = _context14.sent;
              this.checkoutDialogRenderer = FronteggAdminPortal.renderCheckoutDialog(this.checkoutDialogEl, Object.assign({
                injector: this
              }, opts, {
                options: this.options
              }));
            case 4:
            case "end":
              return _context14.stop();
          }
        }, _callee14, this);
      }));
      function showCheckoutDialog(_x20) {
        return _showCheckoutDialog.apply(this, arguments);
      }
      return showCheckoutDialog;
    }()
  }, {
    key: "hideCheckoutDialog",
    value: function hideCheckoutDialog() {
      try {
        var _this$checkoutDialogR;
        (_this$checkoutDialogR = this.checkoutDialogRenderer) == null ? void 0 : _this$checkoutDialogR.unmount();
      } catch (e) {
        console.error('Failed to unmount checkout dialog', e);
      }
    }

    /**
     * @returns entitlements store data
     */
  }, {
    key: "getFeatureEntitlements",
    value:
    /**
     @param key feature key
     @param customAttributes user attributes
     @returns if the user is entitled to the given feature. Attaching the justification if not
     @throws when entitlement is not enabled via frontegg options
     */
    function getFeatureEntitlements(key, customAttributes) {
      var _this$getEntitlements = this.getEntitlementsQueryData(customAttributes),
        entitlements = _this$getEntitlements.entitlements,
        attributes = _this$getEntitlements.attributes;
      return (0, FronteggReduxStore.getFeatureEntitlements)(entitlements, key, attributes);
    }

    /**
     @param key permission key
     @param customAttributes user attributes
     @returns if the user is entitled to the given permission. Attaching the justification if not
     @throws when entitlement is not enabled via frontegg options
     */
  }, {
    key: "getPermissionEntitlements",
    value: function getPermissionEntitlements(key, customAttributes) {
      var _this$getEntitlements2 = this.getEntitlementsQueryData(customAttributes),
        entitlements = _this$getEntitlements2.entitlements,
        attributes = _this$getEntitlements2.attributes;
      return (0, FronteggReduxStore.getPermissionEntitlements)(entitlements, key, attributes);
    }

    /**
     @param options - including permission or feature key
     @param customAttributes user attributes
     @returns if the user is entitled to the given permission or feature. Attaching the justification if not
     @throws when entitlement is not enabled via frontegg options
     */
  }, {
    key: "getEntitlements",
    value: function getEntitlements(options, customAttributes) {
      // we use permissionKey (options key) to decide if the user is using permission key or feature key
      if ('permissionKey' in options) {
        return this.getPermissionEntitlements(options.permissionKey, customAttributes);
      }
      return this.getFeatureEntitlements(options.featureKey, customAttributes);
    }

    /**
     * Load entitlements
     * @param callback called on request completed with true if succeeded, false if failed
     */
  }, {
    key: "loadEntitlements",
    value: function loadEntitlements(callback) {
      this.store.dispatch({
        type: 'auth/loadEntitlements',
        payload: {
          callback: callback
        }
      });
    }

    /**
     * Redirects to the step up page with the max age param and set the redirect url in the local storage
     * @param options.maxAge optional max age
     */
  }, {
    key: "stepUp",
    value: function stepUp(options) {
      var state = this.store.getState();
      var _ref12 = (state == null ? void 0 : state.auth) || {},
        hostedLoginBox = _ref12.hostedLoginBox,
        stepUpUrl = _ref12.routes.stepUpUrl;
      if (hostedLoginBox) {
        this.store.dispatch({
          type: 'auth/stepUpHostedLogin',
          payload: options
        });
        return;
      }
      (0, FronteggReduxStore.redirectByStepUpUrl)(stepUpUrl, state.auth.onRedirectTo, options == null ? void 0 : options.maxAge);
    }

    /**
     * @param options.maxAge optional max age
     * @return true when user is stepped up, false otherwise
     */
  }, {
    key: "isSteppedUp",
    value: function isSteppedUp() {
      var _this$store$getState;
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _ref13 = ((_this$store$getState = this.store.getState()) == null ? void 0 : _this$store$getState.auth) || {},
        user = _ref13.user;
      return (0, FronteggReduxStore.isSteppedUp)(user, options);
    }

    /**
     * Switch tenant
     * @param payload.tenantId tenant id to switch for
     * @param payload.callback optional callback to be called with true / false for success / failure tenant switching operation
     */
  }, {
    key: "switchTenant",
    value: function switchTenant(payload) {
      this.store.dispatch({
        type: 'auth/switchTenant',
        payload: payload
      });
    }
  }, {
    key: "close",
    value: function close() {
      this.hideAdminPortal();
      this.hideCheckoutDialog();
    }
  }]);
  return FronteggApp;
}();
exports.FronteggApp = FronteggApp;
