// @frontegg/react-hooks/node/common/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ShadowDomContext: true,
  useShadowDom: true,
  useRootState: true
};
exports.useShadowDom = exports.useRootState = exports.ShadowDomContext = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _react = require("react");
var _FronteggStoreContext = require("../FronteggStoreContext");
var _useSnapshot = require("../useSnapshot");
var _restApi = require("@frontegg/rest-api");
var _CustomComponentHolder = require("./CustomComponentHolder");
Object.keys(_CustomComponentHolder).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _CustomComponentHolder[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _CustomComponentHolder[key];
    }
  });
});
const _excluded = ["urlStrategy", "onRedirectTo", "renderByRoute", "customLoader", "customStyles", "contextOptions"];
const ShadowDomContext = /*#__PURE__*/(0, _react.createContext)({});
exports.ShadowDomContext = ShadowDomContext;
const defaultPathOnRedirectTo = path => {
  var _window$history, _window$history$pushS;
  return typeof window !== 'undefined' && ((_window$history = window.history) == null ? void 0 : (_window$history$pushS = _window$history.pushState) == null ? void 0 : _window$history$pushS.call(_window$history, null, '', path));
};
const defaultHashOnRedirectTo = path => {
  var _window$history2, _window$history2$push;
  return typeof window !== 'undefined' && ((_window$history2 = window.history) == null ? void 0 : (_window$history2$push = _window$history2.pushState) == null ? void 0 : _window$history2$push.call(_window$history2, null, '', `#${path}`));
};
const useShadowDom = () => {
  var _rest$injector$name, _rest$injector;
  const context = (0, _react.useContext)(ShadowDomContext);
  const {
      urlStrategy = 'path',
      onRedirectTo = urlStrategy === 'path' ? defaultPathOnRedirectTo : defaultHashOnRedirectTo,
      renderByRoute = false,
      customLoader = false,
      customStyles,
      contextOptions
    } = context,
    rest = (0, _objectWithoutPropertiesLoose2.default)(context, _excluded);
  const appName = (_rest$injector$name = (_rest$injector = rest.injector) == null ? void 0 : _rest$injector.name) != null ? _rest$injector$name : 'default';
  return (0, _extends2.default)({
    isShadowDom: true,
    onRedirectTo,
    renderByRoute,
    customLoader,
    customStyles,
    urlStrategy,
    contextOptions: (0, _extends2.default)({
      requestCredentials: 'include'
    }, contextOptions)
  }, rest, {
    appName,
    fetchClient: new _restApi.FetchClient(appName),
    apiClient: (0, _restApi.createApiClient)(appName)
  });
};
exports.useShadowDom = useShadowDom;
const useRootState = () => {
  return (0, _useSnapshot.useSnapshot)((0, _FronteggStoreContext.useStore)().store.root);
};
exports.useRootState = useRootState;
