// @frontegg/rest-api/node/applications/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ApplicationsApi = void 0;
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
/**
 * UsersApplicationsApi class for handling applications-related operations.
 */
class ApplicationsApi extends _BaseApiClient.BaseApiClient {
  constructor(appName) {
    super(appName);
    this.getUserApplicationsId = async ({
      userId
    }) => {
      return this.get(`${_constants.urls.identity.applications.v1}/${userId}/apps`);
    };
    this.getUsersApplicationsId = async ({
      userIds
    }) => {
      return this.get(`${_constants.urls.identity.applications.v1}/users-apps`, {
        userIds: userIds.join(',')
      });
    };
    this.getTenantsApplications = async () => {
      return this.get(_constants.urls.applications.tenant.v1);
    };
    this.getApplicationsData = async ({
      appIds,
      includeFreeAccess = true
    }) => {
      return this.get(_constants.urls.applications.v1, {
        ids: appIds.join(','),
        _includeFreeAccess: includeFreeAccess
      });
    };
    this.assignUserToApplications = async body => {
      return this.post(`${_constants.urls.identity.applications.v1}/apps-user`, body);
    };
    this.unassignUserFromApplications = async body => {
      return this.delete(`${_constants.urls.identity.applications.v1}/user-apps`, body);
    };
    this.getUsersOfApplications = async ({
      appIds
    }) => {
      return this.get(`${_constants.urls.identity.applications.v1}/apps-users`, {
        appIds: appIds.join(',')
      });
    };
  }

  /**
   * Get applications id array for a single user by user id.
   * @param {IUserApplicationsIdPayload} param0 - user id to find their applications id array.
   * @returns {Promise<string[]>} all app ids as string[].
   */
}
exports.ApplicationsApi = ApplicationsApi;
var _default = new ApplicationsApi('default');
exports.default = _default;
