// @frontegg/rest-api/auth/enums.js
export let SocialLoginProviders;
(function (SocialLoginProviders) {
  SocialLoginProviders["Microsoft"] = "microsoft";
  SocialLoginProviders["Facebook"] = "facebook";
  SocialLoginProviders["Google"] = "google";
  SocialLoginProviders["Github"] = "github";
  SocialLoginProviders["Slack"] = "slack";
  SocialLoginProviders["Apple"] = "apple";
  SocialLoginProviders["Linkedin"] = "linkedin";
})(SocialLoginProviders || (SocialLoginProviders = {}));
export let AuthStrategyEnum;
(function (AuthStrategyEnum) {
  AuthStrategyEnum["EmailAndPassword"] = "EmailAndPassword";
  AuthStrategyEnum["MagicLink"] = "MagicLink";
  AuthStrategyEnum["Code"] = "Code";
  AuthStrategyEnum["NoLocalAuthentication"] = "NoLocalAuthentication";
  AuthStrategyEnum["SmsCode"] = "SmsCode";
  AuthStrategyEnum["WebAuthN"] = "WebAuthN";
  AuthStrategyEnum["UsernameAndPassword"] = "UsernameAndPassword";
  AuthStrategyEnum["UsernameAndMagicLink"] = "UsernameAndMagicLink";
  AuthStrategyEnum["UsernameAndCode"] = "UsernameAndCode";
  AuthStrategyEnum["UsernameAndSms"] = "UsernameAndSms";
})(AuthStrategyEnum || (AuthStrategyEnum = {}));
export let MachineToMachineAuthStrategy;
(function (MachineToMachineAuthStrategy) {
  MachineToMachineAuthStrategy["ClientCredentials"] = "ClientCredentials";
  MachineToMachineAuthStrategy["AccessToken"] = "AccessToken";
})(MachineToMachineAuthStrategy || (MachineToMachineAuthStrategy = {}));
export let SignUpStrategyEnum;
(function (SignUpStrategyEnum) {
  SignUpStrategyEnum["Email"] = "email";
  SignUpStrategyEnum["PhoneNumber"] = "phoneNumber";
})(SignUpStrategyEnum || (SignUpStrategyEnum = {}));
export let EIdentifierType;
(function (EIdentifierType) {
  EIdentifierType["email"] = "email";
  EIdentifierType["phoneNumber"] = "phoneNumber";
  EIdentifierType["username"] = "username";
})(EIdentifierType || (EIdentifierType = {}));
export let PasswordRecoveryStrategyEnum;
(function (PasswordRecoveryStrategyEnum) {
  PasswordRecoveryStrategyEnum["Email"] = "email";
  PasswordRecoveryStrategyEnum["Sms"] = "sms";
})(PasswordRecoveryStrategyEnum || (PasswordRecoveryStrategyEnum = {}));
