// @frontegg/rest-api/node/roles/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.RolesApi = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
const _excluded = ["roleId"],
  _excluded2 = ["roleId"],
  _excluded3 = ["permissionId"];
class RolesApi extends _BaseApiClient.BaseApiClient {
  constructor(appName) {
    super(appName);
    this.getRoles = async () => {
      return this.get(_constants.urls.identity.roles.v1);
    };
    this.addRole = async body => {
      return this.post(_constants.urls.identity.roles.v1, body);
    };
    this.deleteRole = async ({
      roleId
    }) => {
      return this.delete(`${_constants.urls.identity.roles.v1}/${roleId}`);
    };
    this.updateRole = async _ref => {
      let {
          roleId
        } = _ref,
        body = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
      return this.patch(`${_constants.urls.identity.roles.v1}/${roleId}`, body);
    };
    this.attachPermissionsToRole = async _ref2 => {
      let {
          roleId
        } = _ref2,
        body = (0, _objectWithoutPropertiesLoose2.default)(_ref2, _excluded2);
      return this.put(`${_constants.urls.identity.roles.v1}/${roleId}/permissions`, body);
    };
    this.getPermissions = async () => {
      return this.get(_constants.urls.identity.permissions.v1);
    };
    this.attachPermissionToRoles = async _ref3 => {
      let {
          permissionId
        } = _ref3,
        body = (0, _objectWithoutPropertiesLoose2.default)(_ref3, _excluded3);
      return this.put(`${_constants.urls.identity.permissions.v1}/${permissionId}/roles`, body);
    };
    this.getPermissionCategories = async () => {
      return this.get(`${_constants.urls.identity.permissions.v1}/categories`);
    };
    this.addRoleV2 = async body => {
      return this.post(_constants.urls.identity.roles.v2, body);
    };
  }

  /**
   * Gets tenant related roles
   */
}
exports.RolesApi = RolesApi;
var _default = new RolesApi('default');
exports.default = _default;
