// @frontegg/redux-store/node/mocks/subscriptions-mocks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildSubscriptionsActions = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _configActions = _interopRequireDefault(require("./configActions.mocks"));
var _billingActionsMocks = _interopRequireDefault(require("./billingActions-mocks"));
var _plansActions = _interopRequireDefault(require("./plansActions.mocks"));
var _checkoutActions = _interopRequireDefault(require("./checkoutActions.mocks"));
var _stripeActions = _interopRequireDefault(require("./stripeActions.mocks"));
var _vendorPublicConfigActions = _interopRequireDefault(require("./vendorPublicConfigActions.mocks"));
const buildSubscriptionsActions = (store, api, actions) => {
  const configActions = (0, _configActions.default)(store, api, actions);
  const [billingActions, billingStateActions] = (0, _billingActionsMocks.default)(store, api, actions);
  const plansActions = (0, _plansActions.default)(store, api, actions);
  const checkoutActions = (0, _checkoutActions.default)(store, api, actions);
  const stripeActions = (0, _stripeActions.default)(store, api, actions);
  const vendorPublicConfigActions = (0, _vendorPublicConfigActions.default)(store, api, actions);
  return [(0, _extends2.default)({}, configActions, billingActions, plansActions, checkoutActions, stripeActions, vendorPublicConfigActions), {
    config: configActions,
    billing: billingStateActions,
    plans: plansActions,
    checkout: checkoutActions,
    stripe: stripeActions,
    vendorPublicConfig: vendorPublicConfigActions
  }];
};
exports.buildSubscriptionsActions = buildSubscriptionsActions;
