// @frontegg/redux-store/node/mocks/subscriptions-mocks/plansActions.mocks.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _helpers = require("../helpers");
var _Plans = require("../../subscriptions/Plans");
var _helpers2 = require("../../helpers");
var _dummy = require("./dummy");
var _default = (store, api, actions) => {
  const originalActions = (0, _Plans.buildPlansActions)(store, api, actions);
  const mockedActions = (0, _helpers.mockActionsExpect)(originalActions, ['setPlansState', 'resetPlansState']);
  mockedActions.loadPlans = async () => {
    mockedActions.setPlansState({
      loading: true
    });
    await (0, _helpers2.delay)(500);
    mockedActions.setPlansState({
      fetching: false,
      loading: false,
      plans: _dummy.planResponseMock.map(item => {
        var _item$price, _item$price2;
        return (0, _extends2.default)({}, item, {
          price: ((_item$price = item.price) == null ? void 0 : _item$price.amount) || 0,
          currency: ((_item$price2 = item.price) == null ? void 0 : _item$price2.currency) || 'usd',
          recurringInterval: 'month'
        });
      })
    });
  };
  return mockedActions;
};
exports.default = _default;
