// @frontegg/rest-api/node/connectivity/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ConnectivityApi = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
const _excluded = ["eventKey"],
  _excluded2 = ["eventKey"];
class ConnectivityApi extends _BaseApiClient.BaseApiClient {
  constructor(appName) {
    super(appName);
    this.getSlackConfiguration = () => {
      return this.get(`${_constants.urls.integrations.configurations.v1}/slack`);
    };
    this.getSlackScope = () => {
      return this.get(`${_constants.urls.integrations.configurations.v1}/slack/applications`);
    };
    this.getSlackChannels = () => {
      return this.get(`${_constants.urls.integrations.configurations.v1}/slack/channels`);
    };
    this.postSlackConfiguration = data => {
      if (data.id) {
        return this.patch(`${_constants.urls.integrations.configurations.v1}/slack/subscription/${data.id}`, data);
      }
      return this.post(`${_constants.urls.integrations.configurations.v1}/slack/subscriptions`, data);
    };
    this.deleteSlackConfiguration = data => {
      return this.delete(`${_constants.urls.integrations.configurations.v1}/slack/subscriptions/${data.id}`);
    };
    this.postSlackCode = code => {
      return this.post(`${_constants.urls.integrations.configurations.v1}/slack/applications/registrations`, {
        code
      });
    };
    this.getEmailConfiguration = () => {
      return this.get(`${_constants.urls.integrations.configurations.v1}/emails`);
    };
    this.postEmailConfiguration = _ref => {
      let {
          eventKey
        } = _ref,
        data = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
      return this.post(`${_constants.urls.integrations.configurations.v1}/emails/${eventKey}`, data);
    };
    this.patchEmailConfiguration = ({
      eventKey,
      enabled
    }) => {
      return this.patch(`${_constants.urls.integrations.configurations.v1}/emails/${eventKey}`, {
        enabled
      });
    };
    this.deleteEmailSubscriptions = (eventKey, subscriptionId) => {
      return this.delete(`${_constants.urls.integrations.configurations.v1}/emails/${eventKey}/subscriptions/${subscriptionId}`);
    };
    this.putEmailSubscriptions = (subscriptionId, eventKey, data) => {
      return this.put(`${_constants.urls.integrations.configurations.v1}/emails/${eventKey}/subscriptions/${subscriptionId}`, data);
    };
    this.deleteEmailConfiguration = eventKey => {
      return this.delete(`${_constants.urls.integrations.configurations.v1}/emails/${eventKey}`);
    };
    this.getSMSConfiguration = () => {
      return this.get(`${_constants.urls.integrations.configurations.v2}/sms`);
    };
    this.postSMSConfiguration = _ref2 => {
      let {
          eventKey
        } = _ref2,
        data = (0, _objectWithoutPropertiesLoose2.default)(_ref2, _excluded2);
      return this.post(`${_constants.urls.integrations.configurations.v2}/sms/${eventKey}`, data);
    };
    this.patchSMSConfiguration = ({
      eventKey,
      enabled
    }) => {
      return this.patch(`${_constants.urls.integrations.configurations.v2}/sms/${eventKey}`, {
        enabled
      });
    };
    this.deleteSMSSubscriptions = (eventKey, subscriptionId) => {
      return this.delete(`${_constants.urls.integrations.configurations.v2}/sms/${eventKey}/subscriptions/${subscriptionId}`);
    };
    this.putSMSSubscriptions = (subscriptionId, eventKey, data) => {
      return this.put(`${_constants.urls.integrations.configurations.v2}/sms/${eventKey}/subscriptions/${subscriptionId}`, data);
    };
    this.deleteSMSConfiguration = eventKey => {
      return this.delete(`${_constants.urls.integrations.configurations.v2}/sms/${eventKey}`);
    };
    this.getWebhooksConfigurations = () => {
      return this.get(_constants.urls.webhooks.v1);
    };
    this.postWebhooksConfiguration = data => {
      if (data._id) {
        return this.patch(`${_constants.urls.webhooks.v1}/${data._id}`, data);
      } else {
        return this.post(`${_constants.urls.webhooks.v1}/custom`, data);
      }
    };
    this.deleteWebhooksConfiguration = id => {
      return this.delete(`${_constants.urls.webhooks.v1}/${id}`);
    };
    this.getWebhookLog = (id, offset = 0, limit = 10) => {
      const query = new URLSearchParams({
        id,
        offset: `${offset}`,
        limit: `${limit}`
      });
      return this.get(`${_constants.urls.webhooks.v1}/logs/?${query.toString()}`);
    };
    this.postWebhookTest = data => {
      return this.post(`${_constants.urls.webhooks.v1}/test`, data);
    };
    this.postWebhookRetry = id => {
      return this.post(`${_constants.urls.webhooks.v1}/logs/${id}/retries`);
    };
    this.getCategories = () => {
      return this.get(`${_constants.urls.events.configurations.v1}/categories`);
    };
    this.getChannelMaps = channels => {
      return this.get(_constants.urls.events.configurations.v1, {
        channels
      });
    };
  }

  /** Get Slack configuration */
}
exports.ConnectivityApi = ConnectivityApi;
var _default = new ConnectivityApi('default');
exports.default = _default;
