// @frontegg/redux-store/node/mocks/subscriptions-mocks/billingActions-mocks/invoicesActions.mocks.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _helpers = require("../../helpers");
var _dummy = require("../dummy");
var _helpers2 = require("../../../helpers");
var _Invoices = require("../../../subscriptions/Billing/Invoices");
var _default = (store, api, actions) => {
  const originalActions = (0, _Invoices.buildInvoicesActions)(store, api, actions);
  const mockedActions = (0, _helpers.mockActionsExpect)(originalActions, ['setBillingInvoiceState', 'resetBillingInvoiceState']);
  mockedActions.loadBillingInvoices = async () => {
    mockedActions.setBillingInvoiceState({
      loading: true
    });
    await (0, _helpers2.delay)(500);
    const selectPlanTitle = 'Premium';
    mockedActions.setBillingInvoiceState({
      loading: false,
      fetching: false,
      invoices: _dummy.invoicesMock.map(invoice => (0, _extends2.default)({}, invoice, {
        selectedPlan: selectPlanTitle,
        paymentDate: new Date(Date.parse(invoice.paymentDate)),
        totalAmount: +((invoice.totalAmount || 0) / 100).toFixed(2)
      }))
    });
  };
  mockedActions.loadInvoices = mockedActions.loadBillingInvoices;
  return mockedActions;
};
exports.default = _default;
