// @frontegg/redux-store/node/auth/ResetPhoneNumberState/actions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _restApi = require("@frontegg/rest-api");
var _interfaces = require("./interfaces");
var _helpers = require("../../helpers");
var _state = require("./state");
var _interfaces2 = require("../LoginState/interfaces");
const _excluded = ["callback"],
  _excluded2 = ["callback"],
  _excluded3 = ["callback"];
var _default = (store, api, sharedActions) => {
  const actions = sharedActions;
  const setResetPhoneNumberState = state => {
    Object.assign(store.auth.resetPhoneNumberState, state);
  };
  const resetResetPhoneNumberState = () => {
    (0, _helpers.deepResetState)(store, ['auth', 'resetPhoneNumberState'], _state.initialState);
  };
  const resetPhoneNumber = async payload => {
    const {
        callback
      } = payload,
      body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded);
    setResetPhoneNumberState({
      loading: true
    });
    try {
      const res = await api.auth.resetPhoneNumber(body);
      setResetPhoneNumberState({
        loading: false,
        error: undefined,
        resetPhoneNumberToken: res.resetPhoneNumberToken,
        step: _interfaces.ResetPhoneNumberStep.VerifyResetPhoneNumber
      });
      actions.setLoginState({
        email: body.email
      });
    } catch (e) {
      setResetPhoneNumberState({
        loading: false,
        error: (0, _helpers.errorHandler)(e)
      });
      callback == null ? void 0 : callback(null, e);
    }
  };
  const verifyResetPhoneNumber = async payload => {
    const {
        callback
      } = payload,
      body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded2);
    setResetPhoneNumberState({
      loading: true
    });
    try {
      const res = await api.auth.verifyResetPhoneNumber(body);
      setResetPhoneNumberState({
        loading: false,
        error: undefined,
        changePhoneNumberToken: res.changePhoneNumberToken,
        step: _interfaces.ResetPhoneNumberStep.ChangePhoneNumber
      });
      callback == null ? void 0 : callback(true);
    } catch (e) {
      setResetPhoneNumberState({
        loading: false,
        error: (0, _helpers.errorHandler)(e)
      });
      callback == null ? void 0 : callback(null, e);
    }
  };
  const changePhoneNumber = async payload => {
    const {
        callback
      } = payload,
      body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded3);
    setResetPhoneNumberState({
      loading: true
    });
    try {
      const {
        onRedirectTo,
        routes
      } = store.auth;
      await api.auth.changePhoneNumber({
        phoneNumber: body.phoneNumber,
        changePhoneNumberToken: body.changePhoneNumberToken
      });
      actions.setLoginState({
        step: _interfaces2.LoginStep.loginWithSmsOtc
      });
      await actions.passwordlessPreLogin({
        type: _restApi.AuthStrategyEnum.SmsCode,
        email: body.email,
        recaptchaToken: body.recaptchaToken
      });
      onRedirectTo(routes.loginUrl);
      setResetPhoneNumberState({
        loading: false
      });
      callback == null ? void 0 : callback(true);
    } catch (e) {
      setResetPhoneNumberState({
        loading: false,
        error: (0, _helpers.errorHandler)(e)
      });
      callback == null ? void 0 : callback(null, e);
    }
  };
  return {
    setResetPhoneNumberState,
    resetResetPhoneNumberState,
    resetPhoneNumber,
    verifyResetPhoneNumber,
    changePhoneNumber
  };
};
exports.default = _default;
