// @frontegg/js/HostedLogin/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import { AppHolder } from '../AppHolder';
import { ContextHolder } from '@frontegg/rest-api';
export var HostedLogin = /*#__PURE__*/function () {
  function HostedLogin() {
    _classCallCheck(this, HostedLogin);
  }
  _createClass(HostedLogin, null, [{
    key: "setAuthentication",
    value: function setAuthentication(isAuthenticated, accessToken, user) {
      var appName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'default';
      var app = AppHolder.getInstance(appName);
      var contextHolder = ContextHolder["for"](appName);
      if (isAuthenticated && accessToken) {
        contextHolder.setAccessToken(accessToken);
        contextHolder.setUser(Object.assign({}, user, {
          accessToken: accessToken
        }));
        app.store.actions.setAuthState({
          isLoading: false,
          isAuthenticated: isAuthenticated,
          user: user
        });
        app.store.actions.loadTenants();
      } else {
        contextHolder.setAccessToken(null);
        contextHolder.setUser(null);
        app.store.actions.setAuthState({
          isLoading: false,
          isAuthenticated: false,
          user: null
        });
      }
    }
  }]);
  return HostedLogin;
}();
