// @frontegg/redux-store/node/auth/Security/SecurityCenterState/state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initialState = exports.default = void 0;
var _proxy = require("../../../toolkit/proxy");
const initialState = {
  loaders: {},
  errors: {},
  recommendations: undefined,
  insights: undefined,
  hasSecurityChanges: false,
  score: 0,
  breachedPasswordUsersTable: {
    users: [],
    usersPageOffset: 0,
    totalUsersItems: 0,
    totalUsersPages: 0,
    queryParams: {
      _offset: 0,
      _limit: 10
    }
  },
  unenrolledMfaUsersTable: {
    users: [],
    usersPageOffset: 0,
    totalUsersItems: 0,
    totalUsersPages: 0,
    queryParams: {
      _offset: 0,
      _limit: 10
    }
  },
  inactiveUsersTable: {
    users: [],
    usersPageOffset: 0,
    totalUsersItems: 0,
    totalUsersPages: 0,
    queryParams: {
      _offset: 0,
      _limit: 10,
      _maxInactiveSeconds: 7776000
    }
  }
};
exports.initialState = initialState;
var _default = overrideState => (0, _proxy.createProxy)(initialState, overrideState);
exports.default = _default;
