// @frontegg/redux-store/node/subscriptions/Billing/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createBillingState = exports.buildBillingActions = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _Information = require("./Information");
var _Invoices = require("./Invoices");
var _PaymentMethod = require("./PaymentMethod");
var _Subscription = require("./Subscription");
var _proxy = require("../../toolkit/proxy");
const createBillingState = overrideState => {
  return (0, _proxy.createProxy)({
    information: (0, _Information.createInformationState)(overrideState == null ? void 0 : overrideState.information),
    invoices: (0, _Invoices.createInvoicesState)(overrideState == null ? void 0 : overrideState.invoices),
    paymentMethod: (0, _PaymentMethod.createPaymentMethodState)(overrideState == null ? void 0 : overrideState.paymentMethod),
    subscription: (0, _Subscription.createSubscriptionState)(overrideState == null ? void 0 : overrideState.subscription)
  });
};
exports.createBillingState = createBillingState;
const buildBillingActions = (store, api, actions) => {
  const informationActions = (0, _Information.buildInformationActions)(store, api, actions);
  const paymentMethodActions = (0, _PaymentMethod.buildPaymentMethodActions)(store, api, actions);
  const invoicesActions = (0, _Invoices.buildInvoicesActions)(store, api, actions);
  const subscriptionActions = (0, _Subscription.buildSubscriptionActions)(store, api, actions);
  return [(0, _extends2.default)({}, informationActions, paymentMethodActions, invoicesActions, subscriptionActions), {
    information: informationActions,
    invoices: invoicesActions,
    paymentMethod: paymentMethodActions,
    subscription: subscriptionActions
  }];
};
exports.buildBillingActions = buildBillingActions;
