// @frontegg/redux-store/node/auth/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createAuthState: true,
  buildAuthActions: true
};
exports.createAuthState = exports.buildAuthActions = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _AcceptInvitationState = require("./AcceptInvitationState");
var _AccountSettingsState = require("./AccountSettingsState");
var _ActivateAccountState = require("./ActivateAccountState");
var _UnlockAccountState = require("./UnlockAccountState");
var _ApiTokensState = require("./ApiTokensState");
var _ApplicationsState = require("./ApplicationsState");
var _CustomLoginState = require("./CustomLoginState");
var _ForgotPasswordState = require("./ForgotPasswordState");
var _PasswordRotationState = require("./PasswordRotationState");
var _GroupsState = require("./GroupsState");
var _GroupsDialogsState = require("./GroupsDialogsState");
var _ImpersonateState = require("./ImpersonateState");
var _LoginState = require("./LoginState");
var _MfaState = require("./MfaState");
var _MSP = require("./MSP");
var _PasskeysState = require("./PasskeysState");
var _ProfileState = require("./ProfileState");
var _ProvisioningState = require("./ProvisioningState");
var _ResetPhoneNumberState = require("./ResetPhoneNumberState");
var _RolesState = require("./RolesState");
var _RestrictionsState = require("./Security/RestrictionsState");
var _SecurityCenterState = require("./Security/SecurityCenterState");
var _SecurityPolicyState = require("./Security/SecurityPolicyState");
var _SessionsPolicyState = require("./Security/SessionsPolicyState");
var _SessionsState = require("./SessionsState");
var _SignUpState = require("./SignUpState");
var _SmsState = require("./SmsState");
var _SocialLoginState = require("./SocialLoginState");
var _SSOState = require("./SSOState");
var _StepUpState = require("./StepUpState");
var _TeamState = require("./TeamState");
var _TenantsState = require("./TenantsState");
var _UsernamesState = require("./UsernamesState");
var _UsersEmailsPolicyState = require("./UsersEmailsPolicyState");
var _consts = require("./LoginState/consts");
var _helpers = require("../helpers");
var _Entitlements = require("./Entitlements");
var _proxy = require("../toolkit/proxy");
var _interfaces = require("./AcceptInvitationState/interfaces");
Object.keys(_interfaces).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces[key];
    }
  });
});
var _interfaces2 = require("./AccountSettingsState/interfaces");
Object.keys(_interfaces2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces2[key];
    }
  });
});
var _interfaces3 = require("./ActivateAccountState/interfaces");
Object.keys(_interfaces3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces3[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces3[key];
    }
  });
});
var _interfaces4 = require("./UnlockAccountState/interfaces");
Object.keys(_interfaces4).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces4[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces4[key];
    }
  });
});
var _interfaces5 = require("./ApiTokensState/interfaces");
Object.keys(_interfaces5).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces5[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces5[key];
    }
  });
});
var _interfaces6 = require("./ApplicationsState/interfaces");
Object.keys(_interfaces6).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces6[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces6[key];
    }
  });
});
var _interfaces7 = require("./CustomLoginState/interfaces");
Object.keys(_interfaces7).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces7[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces7[key];
    }
  });
});
var _interfaces8 = require("./Entitlements/interfaces");
Object.keys(_interfaces8).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces8[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces8[key];
    }
  });
});
var _helpers2 = require("./Entitlements/helpers");
Object.keys(_helpers2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _helpers2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _helpers2[key];
    }
  });
});
var _interfaces9 = require("./ForgotPasswordState/interfaces");
Object.keys(_interfaces9).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces9[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces9[key];
    }
  });
});
var _interfaces10 = require("./PasswordRotationState/interfaces");
Object.keys(_interfaces10).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces10[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces10[key];
    }
  });
});
var _interfaces11 = require("./GroupsState/interfaces");
Object.keys(_interfaces11).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces11[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces11[key];
    }
  });
});
var _interfaces12 = require("./GroupsDialogsState/interfaces");
Object.keys(_interfaces12).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces12[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces12[key];
    }
  });
});
var _interfaces13 = require("./ImpersonateState/interfaces");
Object.keys(_interfaces13).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces13[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces13[key];
    }
  });
});
var _interfaces14 = require("./LoginState/interfaces");
Object.keys(_interfaces14).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces14[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces14[key];
    }
  });
});
var _interfaces15 = require("./MfaState/interfaces");
Object.keys(_interfaces15).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces15[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces15[key];
    }
  });
});
var _interfaces16 = require("./MSP/interfaces");
Object.keys(_interfaces16).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces16[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces16[key];
    }
  });
});
var _interfaces17 = require("./PasskeysState/interfaces");
Object.keys(_interfaces17).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces17[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces17[key];
    }
  });
});
var _interfaces18 = require("./ProfileState/interfaces");
Object.keys(_interfaces18).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces18[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces18[key];
    }
  });
});
var _interfaces19 = require("./ProvisioningState/interfaces");
Object.keys(_interfaces19).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces19[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces19[key];
    }
  });
});
var _interfaces20 = require("./ResetPhoneNumberState/interfaces");
Object.keys(_interfaces20).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces20[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces20[key];
    }
  });
});
var _interfaces21 = require("./RolesState/interfaces");
Object.keys(_interfaces21).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces21[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces21[key];
    }
  });
});
var _interfaces22 = require("./Security/RestrictionsState/interfaces");
Object.keys(_interfaces22).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces22[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces22[key];
    }
  });
});
var _interfaces23 = require("./Security/SecurityCenterState/interfaces");
Object.keys(_interfaces23).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces23[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces23[key];
    }
  });
});
var _interfaces24 = require("./Security/SecurityPolicyState/interfaces");
Object.keys(_interfaces24).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces24[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces24[key];
    }
  });
});
var _interfaces25 = require("./Security/SessionsPolicyState/interfaces");
Object.keys(_interfaces25).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces25[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces25[key];
    }
  });
});
var _interfaces26 = require("./SessionsState/interfaces");
Object.keys(_interfaces26).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces26[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces26[key];
    }
  });
});
var _interfaces27 = require("./SignUpState/interfaces");
Object.keys(_interfaces27).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces27[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces27[key];
    }
  });
});
var _interfaces28 = require("./SmsState/interfaces");
Object.keys(_interfaces28).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces28[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces28[key];
    }
  });
});
var _interfaces29 = require("./SocialLoginState/interfaces");
Object.keys(_interfaces29).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces29[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces29[key];
    }
  });
});
var _interfaces30 = require("./SSOState/interfaces");
Object.keys(_interfaces30).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces30[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces30[key];
    }
  });
});
var _interfaces31 = require("./StepUpState/interfaces");
Object.keys(_interfaces31).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces31[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces31[key];
    }
  });
});
var _interfaces32 = require("./TeamState/interfaces");
Object.keys(_interfaces32).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces32[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces32[key];
    }
  });
});
var _interfaces33 = require("./TenantsState/interfaces");
Object.keys(_interfaces33).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces33[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces33[key];
    }
  });
});
var _interfaces34 = require("./UsernamesState/interfaces");
Object.keys(_interfaces34).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces34[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces34[key];
    }
  });
});
var _interfaces35 = require("./UsersEmailsPolicyState/interfaces");
Object.keys(_interfaces35).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces35[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces35[key];
    }
  });
});
var _interfaces36 = require("./interfaces");
Object.keys(_interfaces36).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces36[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces36[key];
    }
  });
});
const _excluded = ["routes"],
  _excluded2 = ["requestName"];
const createAuthState = _overrideState => {
  const _ref = _overrideState != null ? _overrideState : {},
    {
      routes
    } = _ref,
    overrideState = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  return (0, _proxy.createProxy)((0, _extends2.default)({
    routes: (0, _extends2.default)({}, _consts.defaultFronteggRoutes, routes),
    onRedirectTo: () => void 0,
    isAuthenticated: false,
    isLoading: true,
    isSSOAuth: false,
    user: null,
    error: undefined,
    errorByRequest: {},
    userIp: undefined,
    userIpData: {
      loading: false
    }
  }, overrideState, {
    // nested states
    acceptInvitationState: (0, _AcceptInvitationState.createAcceptInvitationState)(overrideState == null ? void 0 : overrideState.acceptInvitationState),
    accountSettingsState: (0, _AccountSettingsState.createAccountSettingsState)(overrideState == null ? void 0 : overrideState.accountSettingsState),
    activateAccountState: (0, _ActivateAccountState.createActivateAccountState)(overrideState == null ? void 0 : overrideState.activateAccountState),
    unlockAccountState: (0, _UnlockAccountState.createUnlockAccountState)(overrideState == null ? void 0 : overrideState.unlockAccountState),
    apiTokensState: (0, _ApiTokensState.createApiTokensState)(overrideState == null ? void 0 : overrideState.apiTokensState),
    applicationsState: (0, _ApplicationsState.createApplicationsState)(overrideState == null ? void 0 : overrideState.applicationsState),
    customLoginState: (0, _CustomLoginState.createCustomLoginState)(overrideState == null ? void 0 : overrideState.customLoginState),
    forgotPasswordState: (0, _ForgotPasswordState.createForgotPasswordState)(overrideState == null ? void 0 : overrideState.forgotPasswordState),
    passwordRotationState: (0, _PasswordRotationState.createPasswordRotationState)(overrideState == null ? void 0 : overrideState.passwordRotationState),
    groupsState: (0, _GroupsState.createGroupsState)(overrideState == null ? void 0 : overrideState.groupsState),
    groupsDialogsState: (0, _GroupsDialogsState.createGroupsDialogsState)(overrideState == null ? void 0 : overrideState.groupsDialogsState),
    impersonateState: (0, _ImpersonateState.createImpersonateState)(overrideState == null ? void 0 : overrideState.impersonateState),
    loginState: (0, _LoginState.createLoginState)(overrideState == null ? void 0 : overrideState.loginState),
    mfaState: (0, _MfaState.createMfaState)(overrideState == null ? void 0 : overrideState.mfaState),
    allAccountsState: (0, _MSP.createAllAccountsState)(overrideState == null ? void 0 : overrideState.allAccountsState),
    allAccountsDialogsState: (0, _MSP.createAllAccountsDialogState)(overrideState == null ? void 0 : overrideState.allAccountsDialogsState),
    passkeysState: (0, _PasskeysState.createPasskeysState)(overrideState == null ? void 0 : overrideState.passkeysState),
    profileState: (0, _ProfileState.createProfileState)(overrideState == null ? void 0 : overrideState.profileState),
    provisioningState: (0, _ProvisioningState.createProvisioningState)(overrideState == null ? void 0 : overrideState.provisioningState),
    resetPhoneNumberState: (0, _ResetPhoneNumberState.createResetPhoneNumberState)(overrideState == null ? void 0 : overrideState.resetPhoneNumberState),
    rolesState: (0, _RolesState.createRolesState)(overrideState == null ? void 0 : overrideState.rolesState),
    restrictionsState: (0, _RestrictionsState.createRestrictionsState)(overrideState == null ? void 0 : overrideState.restrictionsState),
    securityCenterState: (0, _SecurityCenterState.createSecurityCenterState)(overrideState == null ? void 0 : overrideState.securityCenterState),
    securityPolicyState: (0, _SecurityPolicyState.createSecurityPolicyState)(overrideState == null ? void 0 : overrideState.securityPolicyState),
    sessionsPolicyState: (0, _SessionsPolicyState.createSessionsPolicyState)(overrideState == null ? void 0 : overrideState.sessionsPolicyState),
    sessionsState: (0, _SessionsState.createSessionsState)(overrideState == null ? void 0 : overrideState.sessionsState),
    signUpState: (0, _SignUpState.createSignUpState)(overrideState == null ? void 0 : overrideState.signUpState),
    smsState: (0, _SmsState.createSmsState)(overrideState == null ? void 0 : overrideState.smsState),
    socialLoginState: (0, _SocialLoginState.createSocialLoginState)(overrideState == null ? void 0 : overrideState.socialLoginState),
    ssoState: (0, _SSOState.createSSOState)(overrideState == null ? void 0 : overrideState.ssoState),
    stepUpState: (0, _StepUpState.createStepUpState)(overrideState == null ? void 0 : overrideState.stepUpState),
    teamState: (0, _TeamState.createTeamState)(overrideState == null ? void 0 : overrideState.teamState),
    tenantsState: (0, _TenantsState.createTenantsState)(overrideState == null ? void 0 : overrideState.tenantsState),
    usernamesState: (0, _UsernamesState.createUsernamesState)(overrideState == null ? void 0 : overrideState.usernamesState),
    userEmailPolicyState: (0, _UsersEmailsPolicyState.createUserEmailPolicyState)(overrideState == null ? void 0 : overrideState.userEmailPolicyState)
  }));
};
exports.createAuthState = createAuthState;
const buildAuthActions = (store, api, actions, snapshotAuthState) => {
  const setAuthState = state => {
    Object.keys(state).forEach(key => {
      const authKey = key;
      if ((0, _helpers.isProxy)(store.auth[authKey])) {
        Object.assign(store.auth[authKey], state[authKey]);
      } else {
        store.auth[authKey] = state[authKey];
      }
    });
  };
  const resetAuthState = (state = {}) => {
    (0, _helpers.deepResetState)(store, ['auth'], (0, _extends2.default)({}, snapshotAuthState, state));
  };
  const setUser = user => {
    setAuthState({
      user
    });
  };
  const setErrorByRequestName = _ref2 => {
    let {
        requestName
      } = _ref2,
      data = (0, _objectWithoutPropertiesLoose2.default)(_ref2, _excluded2);
    store.auth.errorByRequest = (0, _extends2.default)({}, store.auth.errorByRequest, {
      [requestName]: (0, _extends2.default)({}, data, {
        isError: true
      })
    });
  };
  const acceptInvitationActions = (0, _AcceptInvitationState.buildAcceptInvitationActions)(store, api, actions);
  const accountSettingsActions = (0, _AccountSettingsState.buildAccountSettingsActions)(store, api, actions);
  const activateAccountActions = (0, _ActivateAccountState.buildActivateAccountActions)(store, api, actions);
  const unlockAccountActions = (0, _UnlockAccountState.buildUnlockAccountActions)(store, api, actions);
  const apiTokensActions = (0, _ApiTokensState.buildApiTokensActions)(store, api, actions);
  const applicationsActions = (0, _ApplicationsState.buildApplicationsActions)(store, api, actions);
  const customLoginActions = (0, _CustomLoginState.buildCustomLoginActions)(store, api, actions);
  const entitlementsActions = (0, _Entitlements.buildEntitlementsActions)(store, api, actions);
  const forgotPasswordActions = (0, _ForgotPasswordState.buildForgotPasswordActions)(store, api, actions);
  const passwordRotationActions = (0, _PasswordRotationState.buildPasswordRotationActions)(store, api, actions);
  const groupsActions = (0, _GroupsState.buildGroupsActions)(store, api, actions);
  const groupsDialogsActions = (0, _GroupsDialogsState.buildGroupsDialogsActions)(store, api, actions);
  const impersonateActions = (0, _ImpersonateState.buildImpersonateActions)(store, api, actions);
  const loginActions = (0, _LoginState.buildLoginActions)(store, api, actions);
  const mfaActions = (0, _MfaState.buildMfaActions)(store, api, actions);
  const allAccountsActions = (0, _MSP.buildAllAccountsActions)(store, api, actions);
  const allAccountsDialogActions = (0, _MSP.buildAllAccountsDialogActions)(store, api, actions);
  const passkeysActions = (0, _PasskeysState.buildPasskeysActions)(store, api, actions);
  const profileActions = (0, _ProfileState.buildProfileActions)(store, api, actions);
  const provisioningActions = (0, _ProvisioningState.buildProvisioningActions)(store, api, actions);
  const resetPhoneNumberActions = (0, _ResetPhoneNumberState.buildResetPhoneNumberActions)(store, api, actions);
  const rolesActions = (0, _RolesState.buildRolesActions)(store, api, actions);
  const restrictionsActions = (0, _RestrictionsState.buildRestrictionsActions)(store, api, actions);
  const securityCenterActions = (0, _SecurityCenterState.buildSecurityCenterActions)(store, api, actions);
  const securityPolicyActions = (0, _SecurityPolicyState.buildSecurityPolicyActions)(store, api, actions);
  const sessionsPolicyActions = (0, _SessionsPolicyState.buildSessionsPolicyActions)(store, api, actions);
  const sessionsActions = (0, _SessionsState.buildSessionsActions)(store, api, actions);
  const signUpActions = (0, _SignUpState.buildSignUpActions)(store, api, actions);
  const smsActions = (0, _SmsState.buildSmsActions)(store, api, actions);
  const socialLoginActions = (0, _SocialLoginState.buildSocialLoginActions)(store, api, actions);
  const ssoActions = (0, _SSOState.buildSSOActions)(store, api, actions);
  const stepUpActions = (0, _StepUpState.buildStepUpActions)(store, api, actions);
  const teamActions = (0, _TeamState.buildTeamActions)(store, api, actions);
  const tenantsActions = (0, _TenantsState.buildTenantsActions)(store, api, actions);
  const usernamesActions = (0, _UsernamesState.buildUsernamesActions)(store, api, actions);
  const usersEmailsPolicyActions = (0, _UsersEmailsPolicyState.buildUserEmailPolicyActions)(store, api, actions);
  const stateActions = {
    acceptInvitationActions,
    accountSettingsActions,
    activateAccountActions,
    unlockAccountActions,
    apiTokensActions,
    applicationsActions,
    customLoginActions,
    entitlementsActions,
    forgotPasswordActions,
    passwordRotationActions,
    groupsActions,
    groupsDialogsActions,
    impersonateActions,
    loginActions,
    mfaActions,
    allAccountsActions,
    allAccountsDialogActions,
    passkeysActions,
    profileActions,
    provisioningActions,
    resetPhoneNumberActions,
    rolesActions,
    restrictionsActions,
    securityCenterActions,
    securityPolicyActions,
    sessionsPolicyActions,
    sessionsActions,
    signUpActions,
    smsActions,
    socialLoginActions,
    ssoActions,
    stepUpActions,
    teamActions,
    tenantsActions,
    usernamesActions,
    usersEmailsPolicyActions
  };
  return [(0, _extends2.default)({
    setAuthState,
    setErrorByRequestName,
    /** @deprecated use setAuthState instead */
    setState: setAuthState,
    resetAuthState,
    setUser
  }, Object.values(stateActions).reduce((acc, actions) => (0, _extends2.default)({}, acc, actions), {})), stateActions];
};
exports.buildAuthActions = buildAuthActions;
