// @frontegg/redux-store/node/mocks/auth-mocks/forgotPasswordActions.mocks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _helpers = require("../../helpers");
var _ForgotPasswordState = require("../../auth/ForgotPasswordState");
var _helpers2 = require("../helpers");
var _default = (store, api, actions) => {
  const originalActions = (0, _ForgotPasswordState.buildForgotPasswordActions)(store, api, actions);
  const mockedActions = (0, _helpers2.mockActionsExpect)(originalActions, ['setForgotPasswordState', 'resetForgotPasswordState']);
  mockedActions.loadPasswordConfig = async payload => {
    mockedActions.setForgotPasswordState({
      loading: true
    });
    try {
      mockedActions.setForgotPasswordState({
        loading: false,
        passwordConfig: {
          allowPassphrases: true,
          maxLength: 128,
          minLength: 6,
          minPhraseLength: 6,
          minOptionalTestsToPass: 1
        }
      });
    } catch (e) {
      mockedActions.setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e)
      });
    }
  };
  return mockedActions;
};
exports.default = _default;
