// @frontegg/rest-api/subscriptions/managedSubscriptions.js
import { urls } from '../constants';
import { BaseApiClient } from '../BaseApiClient';
export class ManagedSubscriptionApi extends BaseApiClient {
  constructor(...args) {
    super(...args);
    this.getManagedSubscription = async subscriptionId => {
      return this.get(`${urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}`);
    };
    this.getManagedSubscriptions = async () => {
      return this.get(urls.subscriptions.managedSubscriptions.v1);
    };
    this.cancelManagedSubscription = async subscriptionId => {
      return this.put(`${urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}/cancellations/`, {});
    };
    this.renewManagedSubscription = async subscriptionId => {
      return this.put(`${urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}/renewals/`, {});
    };
    this.updateManagedSubscription = async (subscriptionId, {
      paymentMethodId,
      planId
    }) => {
      return this.put(`${urls.subscriptions.managedSubscriptions.v1}/${subscriptionId}`, {
        paymentMethodId,
        planId
      });
    };
  }
}
