// @frontegg/rest-api/node/index.js
/** @license Frontegg v7.79.0
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  api: true,
  createApiClient: true,
  AuthStrategyEnum: true,
  EIdentifierType: true,
  MachineToMachineAuthStrategy: true,
  SocialLoginProviders: true,
  setTabTenantInSessionStorage: true,
  getTabTenantFromSessionStorage: true,
  removeTabTenantFromSessionStorage: true,
  getCurrentUserTenantsFunction: true,
  FRONTEGG_SEPARATE_TABS_BY_TENANT: true,
  ISubscriptionCancellationPolicy: true,
  ISubscriptionStatus: true,
  PaymentMethodType: true,
  ProviderType: true,
  FronteggApiError: true,
  fetch: true,
  FetchClient: true,
  ContextHolder: true,
  FronteggContext: true,
  GENERIC_ERROR_CODE: true,
  GENERIC_ERROR_MESSAGE: true
};
Object.defineProperty(exports, "AuthStrategyEnum", {
  enumerable: true,
  get: function () {
    return _auth.AuthStrategyEnum;
  }
});
Object.defineProperty(exports, "ContextHolder", {
  enumerable: true,
  get: function () {
    return _ContextHolder.ContextHolder;
  }
});
Object.defineProperty(exports, "EIdentifierType", {
  enumerable: true,
  get: function () {
    return _auth.EIdentifierType;
  }
});
Object.defineProperty(exports, "FRONTEGG_SEPARATE_TABS_BY_TENANT", {
  enumerable: true,
  get: function () {
    return _auth.FRONTEGG_SEPARATE_TABS_BY_TENANT;
  }
});
Object.defineProperty(exports, "FetchClient", {
  enumerable: true,
  get: function () {
    return _FetchClient.FetchClient;
  }
});
Object.defineProperty(exports, "FronteggApiError", {
  enumerable: true,
  get: function () {
    return _error.FronteggApiError;
  }
});
Object.defineProperty(exports, "FronteggContext", {
  enumerable: true,
  get: function () {
    return _ContextHolder.FronteggContext;
  }
});
Object.defineProperty(exports, "GENERIC_ERROR_CODE", {
  enumerable: true,
  get: function () {
    return _constants.GENERIC_ERROR_CODE;
  }
});
Object.defineProperty(exports, "GENERIC_ERROR_MESSAGE", {
  enumerable: true,
  get: function () {
    return _constants.GENERIC_ERROR_MESSAGE;
  }
});
Object.defineProperty(exports, "ISubscriptionCancellationPolicy", {
  enumerable: true,
  get: function () {
    return _subscriptions.ISubscriptionCancellationPolicy;
  }
});
Object.defineProperty(exports, "ISubscriptionStatus", {
  enumerable: true,
  get: function () {
    return _subscriptions.ISubscriptionStatus;
  }
});
Object.defineProperty(exports, "MachineToMachineAuthStrategy", {
  enumerable: true,
  get: function () {
    return _auth.MachineToMachineAuthStrategy;
  }
});
Object.defineProperty(exports, "PaymentMethodType", {
  enumerable: true,
  get: function () {
    return _subscriptions.PaymentMethodType;
  }
});
Object.defineProperty(exports, "ProviderType", {
  enumerable: true,
  get: function () {
    return _subscriptions.ProviderType;
  }
});
Object.defineProperty(exports, "SocialLoginProviders", {
  enumerable: true,
  get: function () {
    return _auth.SocialLoginProviders;
  }
});
exports.default = exports.createApiClient = exports.api = void 0;
Object.defineProperty(exports, "fetch", {
  enumerable: true,
  get: function () {
    return _FetchClient.default;
  }
});
Object.defineProperty(exports, "getCurrentUserTenantsFunction", {
  enumerable: true,
  get: function () {
    return _auth.getCurrentUserTenantsFunction;
  }
});
Object.defineProperty(exports, "getTabTenantFromSessionStorage", {
  enumerable: true,
  get: function () {
    return _auth.getTabTenantFromSessionStorage;
  }
});
Object.defineProperty(exports, "removeTabTenantFromSessionStorage", {
  enumerable: true,
  get: function () {
    return _auth.removeTabTenantFromSessionStorage;
  }
});
Object.defineProperty(exports, "setTabTenantInSessionStorage", {
  enumerable: true,
  get: function () {
    return _auth.setTabTenantInSessionStorage;
  }
});
var _auth = _interopRequireWildcard(require("./auth"));
var _subscriptions = _interopRequireWildcard(require("./subscriptions"));
var _error = require("./error");
var _FetchClient = _interopRequireWildcard(require("./FetchClient"));
var _ContextHolder = require("./ContextHolder");
var _secutiryPoilicy = _interopRequireWildcard(require("./auth/secutiry-poilicy"));
var _users = _interopRequireWildcard(require("./users"));
var _teams = _interopRequireWildcard(require("./teams"));
var _metadata = _interopRequireWildcard(require("./metadata"));
var _reports = _interopRequireWildcard(require("./reports"));
var _connectivity = _interopRequireWildcard(require("./connectivity"));
var _notifications = _interopRequireWildcard(require("./notifications"));
var _audits = _interopRequireWildcard(require("./audits"));
var _tenants = _interopRequireWildcard(require("./tenants"));
var _accountSettings = _interopRequireWildcard(require("./account-settings"));
var _roles = _interopRequireWildcard(require("./roles"));
var _vendor = _interopRequireWildcard(require("./vendor"));
var _subTenants = _interopRequireWildcard(require("./sub-tenants"));
var _featureFlags = _interopRequireWildcard(require("./feature-flags"));
var _directory = _interopRequireWildcard(require("./directory"));
var _impersonate = _interopRequireWildcard(require("./impersonate"));
var _groups = _interopRequireWildcard(require("./groups"));
var _entitlements = _interopRequireWildcard(require("./entitlements"));
var _securityCenter = _interopRequireWildcard(require("./security-center"));
var _userPhoneNumbers = _interopRequireWildcard(require("./user-phone-numbers"));
var _applications = _interopRequireWildcard(require("./applications"));
var _velo = _interopRequireWildcard(require("./velo"));
var _usernames = _interopRequireWildcard(require("./usernames"));
var _usersEmailsPolicy = _interopRequireWildcard(require("./users-emails-policy"));
var _interfaces = require("./interfaces");
Object.keys(_interfaces).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces[key];
    }
  });
});
var _interfaces2 = require("./auth/interfaces");
Object.keys(_interfaces2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces2[key];
    }
  });
});
var _enums = require("./auth/enums");
Object.keys(_enums).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _enums[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _enums[key];
    }
  });
});
var _interfaces3 = require("./teams/interfaces");
Object.keys(_interfaces3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces3[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces3[key];
    }
  });
});
var _interfaces4 = require("./metadata/interfaces");
Object.keys(_interfaces4).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces4[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces4[key];
    }
  });
});
var _interfaces5 = require("./reports/interfaces");
Object.keys(_interfaces5).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces5[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces5[key];
    }
  });
});
var _interfaces6 = require("./connectivity/interfaces");
Object.keys(_interfaces6).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces6[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces6[key];
    }
  });
});
var _interfaces7 = require("./notifications/interfaces");
Object.keys(_interfaces7).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces7[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces7[key];
    }
  });
});
var _interfaces8 = require("./audits/interfaces");
Object.keys(_interfaces8).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces8[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces8[key];
    }
  });
});
var _interfaces9 = require("./tenants/interfaces");
Object.keys(_interfaces9).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces9[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces9[key];
    }
  });
});
var _interfaces10 = require("./account-settings/interfaces");
Object.keys(_interfaces10).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces10[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces10[key];
    }
  });
});
var _interfaces11 = require("./roles/interfaces");
Object.keys(_interfaces11).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces11[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces11[key];
    }
  });
});
var _interfaces12 = require("./subscriptions/interfaces");
Object.keys(_interfaces12).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces12[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces12[key];
    }
  });
});
var _interfaces13 = require("./vendor/interfaces");
Object.keys(_interfaces13).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces13[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces13[key];
    }
  });
});
var _interfaces14 = require("./sub-tenants/interfaces");
Object.keys(_interfaces14).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces14[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces14[key];
    }
  });
});
var _routers = require("./routers");
Object.keys(_routers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _routers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _routers[key];
    }
  });
});
var _index = require("./feature-flags/index");
Object.keys(_index).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _index[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _index[key];
    }
  });
});
var _interfaces15 = require("./feature-flags/interfaces");
Object.keys(_interfaces15).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces15[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces15[key];
    }
  });
});
var _index2 = require("./directory/index");
Object.keys(_index2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _index2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _index2[key];
    }
  });
});
var _interfaces16 = require("./directory/interfaces");
Object.keys(_interfaces16).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces16[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces16[key];
    }
  });
});
var _interfaces17 = require("./impersonate/interfaces");
Object.keys(_interfaces17).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces17[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces17[key];
    }
  });
});
var _interfaces18 = require("./groups/interfaces");
Object.keys(_interfaces18).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces18[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces18[key];
    }
  });
});
var _enums2 = require("./groups/enums");
Object.keys(_enums2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _enums2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _enums2[key];
    }
  });
});
var _interfaces19 = require("./users/interfaces");
Object.keys(_interfaces19).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces19[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces19[key];
    }
  });
});
var _interfaces20 = require("./entitlements/interfaces");
Object.keys(_interfaces20).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces20[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces20[key];
    }
  });
});
var _interfaces21 = require("./security-center/interfaces");
Object.keys(_interfaces21).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces21[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces21[key];
    }
  });
});
var _interfaces22 = require("./user-phone-numbers/interfaces");
Object.keys(_interfaces22).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces22[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces22[key];
    }
  });
});
var _interfaces23 = require("./applications/interfaces");
Object.keys(_interfaces23).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces23[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces23[key];
    }
  });
});
var _interfaces24 = require("./velo/interfaces");
Object.keys(_interfaces24).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces24[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces24[key];
    }
  });
});
var _interfaces25 = require("./users-emails-policy/interfaces");
Object.keys(_interfaces25).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces25[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces25[key];
    }
  });
});
var _constants = require("./constants");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * Frontegg API client for default app
 *
 * @deprecated use createApiClient instead
 */
const api = {
  auth: _auth.default,
  securityPolicy: _secutiryPoilicy.default,
  teams: _teams.default,
  metadata: _metadata.default,
  reports: _reports.default,
  connectivity: _connectivity.default,
  notifications: _notifications.default,
  audits: _audits.default,
  tenants: _tenants.default,
  accountSettings: _accountSettings.default,
  roles: _roles.default,
  subscriptions: _subscriptions.default,
  vendor: _vendor.default,
  subTenants: _subTenants.default,
  featureFlags: _featureFlags.default,
  directory: _directory.default,
  impersonate: _impersonate.default,
  groups: _groups.default,
  users: _users.default,
  entitlements: _entitlements.default,
  securityCenter: _securityCenter.default,
  userPhoneNumbers: _userPhoneNumbers.default,
  applications: _applications.default,
  velo: _velo.default,
  usernames: _usernames.default,
  userEmailPolicy: _usersEmailsPolicy.default
};
exports.api = api;
const createApiClient = appName => ({
  auth: new _auth.AuthenticationApi(appName),
  securityPolicy: new _secutiryPoilicy.SecurityPolicyApi(appName),
  teams: new _teams.TeamsApi(appName),
  metadata: new _metadata.MetadataApi(appName),
  reports: new _reports.ReportsApi(appName),
  connectivity: new _connectivity.ConnectivityApi(appName),
  notifications: new _notifications.NotificationsApi(appName),
  audits: new _audits.AuditsApi(appName),
  tenants: new _tenants.TenantsApi(appName),
  accountSettings: new _accountSettings.AccountSettingsApi(appName),
  roles: new _roles.RolesApi(appName),
  subscriptions: new _subscriptions.SubscriptionsApi(appName),
  vendor: new _vendor.VendorApi(appName),
  subTenants: new _subTenants.SubTenantsApi(appName),
  featureFlags: new _featureFlags.FeatureFlagsApi(appName),
  directory: new _directory.DirectoryApi(appName),
  impersonate: new _impersonate.ImpersonateApi(appName),
  groups: new _groups.GroupsApi(appName),
  users: new _users.UsersApi(appName),
  entitlements: new _entitlements.EntitlementsApi(appName),
  securityCenter: new _securityCenter.SecurityCenterApi(appName),
  userPhoneNumbers: new _userPhoneNumbers.PhoneNumbersApi(appName),
  applications: new _applications.ApplicationsApi(appName),
  velo: new _velo.VeloApi(appName),
  usernames: new _usernames.UsernamesApi(appName),
  userEmailPolicy: new _usersEmailsPolicy.UserEmailPolicyApi(appName)
});
exports.createApiClient = createApiClient;
var _default = {
  fetch: _FetchClient.default,
  FetchClient: _FetchClient.FetchClient,
  ContextHolder: _ContextHolder.ContextHolder,
  FronteggContext: _ContextHolder.FronteggContext,
  api,
  createApiClient,
  FronteggApiError: _error.FronteggApiError,
  AuthStrategyEnum: _auth.AuthStrategyEnum,
  MachineToMachineAuthStrategy: _auth.MachineToMachineAuthStrategy,
  SocialLoginProviders: _auth.SocialLoginProviders,
  ISubscriptionCancellationPolicy: _subscriptions.ISubscriptionCancellationPolicy,
  ISubscriptionStatus: _subscriptions.ISubscriptionStatus,
  PaymentMethodType: _subscriptions.PaymentMethodType,
  ProviderType: _subscriptions.ProviderType,
  EIdentifierType: _auth.EIdentifierType
};
exports.default = _default;
