// @frontegg/rest-api/node/routers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fronteggUsersUrl = exports.fronteggTenantsV3Url = exports.fronteggTenantsUrl = exports.fronteggSilentRefreshTokenUrl = exports.fronteggRefreshTokenUrl = exports.fronteggEntitlementsV2Url = exports.fronteggAuthApiRoutesRegex = void 0;
var _constants = require("./constants");
const fronteggAuthApiRoutesRegex = [
// Logout urls:
// - `${urls.identity.auth.v1}/logout`,
// - `${urls.oauth.v1}/logout`,
/^(.*)\/logout$/g, '/logout',
// Saml urls:
// - `${urls.identity.auth.v1}/user/saml/postlogin`,
// - `${urls.identity.auth.v2}/user/oidc/postlogin`,
/^\/identity\/resources\/auth\/v[0-9]*\/user\/[^\/]*\/postlogin$/g,
// impersonation url:
/^\/identity\/resources\/impersonation\/v[0-9]$/g,
// Passwordless urls:
// - `${urls.identity.auth.v1}/passwordless/magiclink/postlogin`,
// - `${urls.identity.auth.v1}/passwordless/code/postlogin`,
// - `${urls.identity.auth.v1}/passwordless/smscode/postlogin`,
// - `${urls.identity.auth.v1}/passwordless/${type.toLocaleLowerCase()}/postlogin`,
/^\/identity\/resources\/auth\/v[0-9]*\/passwordless\/[^\/]*\/postlogin$/g,
// Webauthn postlogin urls:
/^\/identity\/resources\/auth\/v[0-9]*\/webauthn\/postlogin$/g,
// General login urls:
// - `${urls.identity.auth.v1}/user`,
// - `${urls.identity.auth.v1}/user/token/refresh`,
// - `${urls.identity.users.v1}`,
// - `${urls.identity.users.v1}/signUp`,
// - `${urls.identity.users.v1}/activate`,
// - `${urls.identity.users.v1}/invitation/accept`,
/^\/identity\/resources\/auth\/v[0-9]*\/user$/g, /^\/identity\/resources\/auth\/v[0-9]*\/user\/token\/refresh$/g, /^\/identity\/resources\/users\/v[0-9]*\/signUp$/g, /^\/identity\/resources\/users\/v[0-9]*\/activate$/g, /^\/identity\/resources\/users\/v[0-9]*\/invitation\/accept$/g,
// - `${urls.identity.auth.v1}/user/mfa/verify`,
// - `${urls.identity.auth.v1}/user/mfa/authenticator/enroll/verify`,
// - `${urls.identity.auth.v1}/user/mfa/authenticator/${deviceId}/verify`,
// - `${urls.identity.auth.v1}/user/mfa/sms/enroll/verify`,
// - `${urls.identity.auth.v1}/user/mfa/sms/${deviceId}/verify`,
// - `${urls.identity.auth.v1}/user/mfa/webauthn/enroll/verify`,
// - `${urls.identity.auth.v1}/user/mfa/webauthn/${deviceId}/verify`,
// - `${urls.identity.auth.v1}/user/mfa/emailcode/verify`,
/^\/identity\/resources\/auth\/v[0-9]*\/user\/mfa\/verify$/g, /^\/identity\/resources\/auth\/v[0-9]*\/user\/mfa\/[^\/]*\/[^\/]*\/verify$/g, /^\/identity\/resources\/auth\/v[0-9]*\/user\/mfa\/[^\/]*\/verify$/g,
// Oauth urls:
// - `${urls.oauth.v1}/token`
// - `${urls.oauth.v1}/authorize/silent`
/^\/oauth\/token$/g, /^\/oauth\/authorize\/silent$/g];
exports.fronteggAuthApiRoutesRegex = fronteggAuthApiRoutesRegex;
const fronteggRefreshTokenUrl = `${_constants.urls.identity.auth.v1}/user/token/refresh`;
exports.fronteggRefreshTokenUrl = fronteggRefreshTokenUrl;
const fronteggSilentRefreshTokenUrl = `${_constants.urls.oauth.v1}/authorize/silent`;
exports.fronteggSilentRefreshTokenUrl = fronteggSilentRefreshTokenUrl;
const fronteggUsersUrl = `${_constants.urls.identity.users.v2}/me`;
exports.fronteggUsersUrl = fronteggUsersUrl;
const fronteggTenantsUrl = `${_constants.urls.identity.users.v2}/me/tenants`;
exports.fronteggTenantsUrl = fronteggTenantsUrl;
const fronteggTenantsV3Url = `${_constants.urls.identity.users.v3}/me/tenants`;
exports.fronteggTenantsV3Url = fronteggTenantsV3Url;
const fronteggEntitlementsV2Url = _constants.urls.entitlements.v2;
exports.fronteggEntitlementsV2Url = fronteggEntitlementsV2Url;
