// @frontegg/rest-api/node/teams/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TeamsApi = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
class TeamsApi extends _BaseApiClient.BaseApiClient {
  constructor(appName) {
    super(appName);
    this.getProfile = async params => {
      return this.get(`${_constants.urls.identity.users.v2}/me`, params != null ? params : {
        addRoles: true
      });
    };
    this.updateEmail = async email => {
      return this.post(`${_constants.urls.identity.users.v1}/email/me`, {
        email
      });
    };
    this.verifyEmail = async (email, code) => {
      return this.post(`${_constants.urls.identity.users.v1}/email/me/verify`, {
        email,
        code
      });
    };
    this.updateProfile = async body => {
      return this.put(`${_constants.urls.identity.users.v2}/me`, body);
    };
    this.updateProfileImage = async body => {
      return this.put(`${_constants.urls.team.profile.v1}/me/image/v1`, body, {
        responseType: 'plain',
        contentType: undefined
      });
    };
    this.changePassword = async body => {
      return this.post(`${_constants.urls.identity.users.v1}/passwords/change`, body);
    };
    this.loadUsers = async params => {
      var _params$sort, _params$sort$0$id, _params$sort2, _params$sort2$, _params$sort3, _params$sort3$;
      const filters = (params.filter || []).reduce((p, n) => (0, _extends2.default)({}, p, {
        [n.id]: n.value
      }), {});
      const sorts = (_params$sort = params.sort) != null && _params$sort.length ? {
        sortBy: (_params$sort$0$id = (_params$sort2 = params.sort) == null ? void 0 : (_params$sort2$ = _params$sort2[0]) == null ? void 0 : _params$sort2$.id) != null ? _params$sort$0$id : 'name',
        sortDirection: (_params$sort3 = params.sort) != null && (_params$sort3$ = _params$sort3[0]) != null && _params$sort3$.desc ? 'desc' : 'asc'
      } : null;
      return this.get(_constants.urls.team.members.v1, (0, _extends2.default)({
        pageOffset: params.pageOffset,
        pageSize: params.pageSize
      }, filters, sorts));
    };
    this.addUser = async body => {
      return this.post(_constants.urls.identity.users.v2, (0, _extends2.default)({}, body, {
        provider: 'local'
      }));
    };
    this.deleteUser = async body => {
      return this.delete(`${_constants.urls.team.members.v1}/${body.userId}`);
    };
    this.updateUser = async body => {
      return this.put(_constants.urls.team.members.v1, body);
    };
    this.updateUserExpirationTime = async body => {
      const {
        userId,
        expirationInSeconds
      } = body;
      return this.put(`${_constants.urls.identity.users.temporary.v1}/${userId}`, {
        expirationInSeconds
      });
    };
    this.setPermanentUser = async userId => {
      return this.delete(`${_constants.urls.identity.users.temporary.v1}/${userId}`);
    };
    this.getTemporaryUserConfiguration = async () => {
      return this.get(_constants.urls.identity.users.temporary.configuration.v1);
    };
    this.createInviteUserLink = async body => {
      return this.post(_constants.urls.identity.tenants.invites.user.v1, (0, _extends2.default)({}, body));
    };
    this.updateInviteUserLink = async body => {
      return this.patch(_constants.urls.identity.tenants.invites.user.v1, (0, _extends2.default)({}, body));
    };
    this.deleteInviteUserLink = async () => {
      return this.delete(_constants.urls.identity.tenants.invites.user.v1);
    };
    this.getInviteUserLink = async () => {
      return this.get(_constants.urls.identity.tenants.invites.user.v1);
    };
    this.getInviteLinkConfiguration = async () => {
      return this.get(_constants.urls.identity.tenants.invites.configuration.v1);
    };
    this.loadAvailableRoles = async () => {
      return this.get(_constants.urls.team.roles.v1);
    };
    this.loadAvailablePermissions = async () => {
      return this.get(_constants.urls.identity.permissions.v1);
    };
    this.loadStats = async () => {
      return this.get(`${_constants.urls.team.stats.v1}/members`);
    };
    this.resendActivationLink = async body => {
      return this.post(`${_constants.urls.team.members.v1}/${body.userId}/resendActivationEmail`, {});
    };
    this.resendInvitationLink = async body => {
      return this.post(_constants.urls.identity.users.v2, (0, _extends2.default)({}, body, {
        provider: 'local'
      }));
    };
    this.resendInvitationLinkToAllTenants = async body => {
      return this.post(_constants.urls.identity.users.invitation.resendAll.v1, (0, _extends2.default)({}, body));
    };
    this.disableUser = async userId => {
      return this.post(`${_constants.urls.identity.tenants.users.v1}/${userId}/disable`);
    };
    this.enableUser = async userId => {
      return this.post(`${_constants.urls.identity.tenants.users.v1}/${userId}/enable`);
    };
  }
}
exports.TeamsApi = TeamsApi;
var _default = new TeamsApi('default');
exports.default = _default;
