// @frontegg/rest-api/node/auth/enums.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SocialLoginProviders = exports.SignUpStrategyEnum = exports.PasswordRecoveryStrategyEnum = exports.MachineToMachineAuthStrategy = exports.EIdentifierType = exports.AuthStrategyEnum = void 0;
let SocialLoginProviders;
exports.SocialLoginProviders = SocialLoginProviders;
(function (SocialLoginProviders) {
  SocialLoginProviders["Microsoft"] = "microsoft";
  SocialLoginProviders["Facebook"] = "facebook";
  SocialLoginProviders["Google"] = "google";
  SocialLoginProviders["Github"] = "github";
  SocialLoginProviders["Slack"] = "slack";
  SocialLoginProviders["Apple"] = "apple";
  SocialLoginProviders["Linkedin"] = "linkedin";
})(SocialLoginProviders || (exports.SocialLoginProviders = SocialLoginProviders = {}));
let AuthStrategyEnum;
exports.AuthStrategyEnum = AuthStrategyEnum;
(function (AuthStrategyEnum) {
  AuthStrategyEnum["EmailAndPassword"] = "EmailAndPassword";
  AuthStrategyEnum["MagicLink"] = "MagicLink";
  AuthStrategyEnum["Code"] = "Code";
  AuthStrategyEnum["NoLocalAuthentication"] = "NoLocalAuthentication";
  AuthStrategyEnum["SmsCode"] = "SmsCode";
  AuthStrategyEnum["WebAuthN"] = "WebAuthN";
  AuthStrategyEnum["UsernameAndPassword"] = "UsernameAndPassword";
  AuthStrategyEnum["UsernameAndMagicLink"] = "UsernameAndMagicLink";
  AuthStrategyEnum["UsernameAndCode"] = "UsernameAndCode";
  AuthStrategyEnum["UsernameAndSms"] = "UsernameAndSms";
})(AuthStrategyEnum || (exports.AuthStrategyEnum = AuthStrategyEnum = {}));
let MachineToMachineAuthStrategy;
exports.MachineToMachineAuthStrategy = MachineToMachineAuthStrategy;
(function (MachineToMachineAuthStrategy) {
  MachineToMachineAuthStrategy["ClientCredentials"] = "ClientCredentials";
  MachineToMachineAuthStrategy["AccessToken"] = "AccessToken";
})(MachineToMachineAuthStrategy || (exports.MachineToMachineAuthStrategy = MachineToMachineAuthStrategy = {}));
let SignUpStrategyEnum;
exports.SignUpStrategyEnum = SignUpStrategyEnum;
(function (SignUpStrategyEnum) {
  SignUpStrategyEnum["Email"] = "email";
  SignUpStrategyEnum["PhoneNumber"] = "phoneNumber";
})(SignUpStrategyEnum || (exports.SignUpStrategyEnum = SignUpStrategyEnum = {}));
let EIdentifierType;
exports.EIdentifierType = EIdentifierType;
(function (EIdentifierType) {
  EIdentifierType["email"] = "email";
  EIdentifierType["phoneNumber"] = "phoneNumber";
  EIdentifierType["username"] = "username";
})(EIdentifierType || (exports.EIdentifierType = EIdentifierType = {}));
let PasswordRecoveryStrategyEnum;
exports.PasswordRecoveryStrategyEnum = PasswordRecoveryStrategyEnum;
(function (PasswordRecoveryStrategyEnum) {
  PasswordRecoveryStrategyEnum["Email"] = "email";
  PasswordRecoveryStrategyEnum["Sms"] = "sms";
})(PasswordRecoveryStrategyEnum || (exports.PasswordRecoveryStrategyEnum = PasswordRecoveryStrategyEnum = {}));
