// @frontegg/rest-api/node/user-phone-numbers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PhoneNumbersApi = void 0;
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
class PhoneNumbersApi extends _BaseApiClient.BaseApiClient {
  constructor(appName) {
    super(appName);
    this.getUserPhoneNumbers = async queryParams => {
      return this.get(`${_constants.urls.identity.phoneNumbers.v1}`, queryParams);
    };
    this.getUserOwnPhoneNumbers = async () => {
      return this.get(`${_constants.urls.identity.phoneNumbers.v1}/me`);
    };
    this.createPhoneNumber = async body => {
      return this.post(`${_constants.urls.identity.phoneNumbers.v1}`, body);
    };
    this.deletePhoneNumber = async phoneId => {
      return this.delete(`${_constants.urls.identity.phoneNumbers.v1}/${phoneId}`);
    };
    this.preVerifyPhoneNumber = async body => {
      return this.post(`${_constants.urls.identity.phoneNumbers.v1}/preverify`, body);
    };
    this.verifyPhoneNumber = async body => {
      return this.post(`${_constants.urls.identity.phoneNumbers.v1}/verify`, body);
    };
    this.verifyDeletePhoneNumber = async (phoneId, body) => {
      return this.post(`${_constants.urls.identity.phoneNumbers.v1}/${phoneId}/delete/verify`, body);
    };
  }

  /**
   * Get phone numbers of user
   */
}
exports.PhoneNumbersApi = PhoneNumbersApi;
var _default = new PhoneNumbersApi('default');
exports.default = _default;
