// @frontegg/js/node/HostedLogin/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HostedLogin = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _AppHolder = require("../AppHolder");
var _restApi = require("@frontegg/rest-api");
var HostedLogin = /*#__PURE__*/function () {
  function HostedLogin() {
    (0, _classCallCheck2["default"])(this, HostedLogin);
  }
  (0, _createClass2["default"])(HostedLogin, null, [{
    key: "setAuthentication",
    value: function setAuthentication(isAuthenticated, accessToken, user) {
      var appName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'default';
      var app = _AppHolder.AppHolder.getInstance(appName);
      var contextHolder = _restApi.ContextHolder["for"](appName);
      if (isAuthenticated && accessToken) {
        contextHolder.setAccessToken(accessToken);
        contextHolder.setUser(Object.assign({}, user, {
          accessToken: accessToken
        }));
        app.store.actions.setAuthState({
          isLoading: false,
          isAuthenticated: isAuthenticated,
          user: user
        });
        app.store.actions.loadTenants();
      } else {
        contextHolder.setAccessToken(null);
        contextHolder.setUser(null);
        app.store.actions.setAuthState({
          isLoading: false,
          isAuthenticated: false,
          user: null
        });
      }
    }
  }]);
  return HostedLogin;
}();
exports.HostedLogin = HostedLogin;
