// @frontegg/rest-api/node/feature-flags/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FeatureFlagsApi = exports.FeatureFlags = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _constants = require("../constants");
var _BaseApiClient = require("../BaseApiClient");
const defaultFeatureFlags = {
  flags: {}
};
class FeatureFlags {
  constructor() {
    var _defaultFeatureFlags$;
    this._flags = {};
    this._flags = (_defaultFeatureFlags$ = defaultFeatureFlags.flags) != null ? _defaultFeatureFlags$ : {};
  }
  static getInstance(name = 'default') {
    return this._instances[name];
  }
  static set(featureFlags, name = 'default') {
    const featureFlagsInstance = new FeatureFlags();
    featureFlagsInstance.set(featureFlags);
    FeatureFlags._instances[name] = featureFlagsInstance;
    return featureFlagsInstance;
  }
  static getFeatureFlags(flags, name) {
    const featureFlagsInstance = this.getInstance(name);
    return flags.map(flag => (featureFlagsInstance == null ? void 0 : featureFlagsInstance._flags[flag]) === 'on');
  }
  get flags() {
    var _this$flags;
    return (_this$flags = this.flags) != null ? _this$flags : {};
  }
  set(featureFlags) {
    var _defaultFeatureFlags$2;
    this._flags = (0, _extends2.default)({}, (_defaultFeatureFlags$2 = defaultFeatureFlags == null ? void 0 : defaultFeatureFlags.flags) != null ? _defaultFeatureFlags$2 : {}, featureFlags != null ? featureFlags : {});
  }
}
exports.FeatureFlags = FeatureFlags;
FeatureFlags._instances = {};
class FeatureFlagsApi extends _BaseApiClient.BaseApiClient {
  constructor(appName) {
    super(appName);
    this.loadFeatureFlags = async () => {
      return this.get(_constants.urls.featureFlags.v1);
    };
  }
}
exports.FeatureFlagsApi = FeatureFlagsApi;
var _default = new FeatureFlagsApi('default');
exports.default = _default;
