// @frontegg/react-hooks/node/auth/entitlements.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePermissionEntitlements = exports.useFeatureEntitlements = exports.useEntitlementsOptions = exports.useEntitlementsActions = exports.useEntitlements = void 0;
var _hooks = require("./hooks");
var _react = require("react");
var _common = require("../common");
var _flags = require("../flags");
var _FronteggStoreContext = require("../FronteggStoreContext");
var _useSnapshot = require("../useSnapshot");
var _reduxStore = require("@frontegg/redux-store");
/**
 * @returns entitlements state as stored in the user
 */
const useEntitlementsState = () => {
  var _useSnapshot$user$ent, _useSnapshot$user;
  const state = (0, _FronteggStoreContext.useStore)().store.auth;
  return (_useSnapshot$user$ent = (_useSnapshot$user = (0, _useSnapshot.useSnapshot)(state).user) == null ? void 0 : _useSnapshot$user.entitlements) != null ? _useSnapshot$user$ent : {};
};

/**
 * @returns user state
 */
const useUserState = () => {
  return (0, _hooks.useAuthUserOrNull)() || undefined;
};

/**
 * @param customAttributes user attributes
 * @returns is entitled query data including entitlement state and final attributes (consumer and frontegg)
 */
const useEntitlementsQueryData = customAttributes => {
  const user = useUserState();
  const entitlements = useEntitlementsState();
  const attributes = {
    custom: customAttributes,
    jwt: user
  };
  return {
    entitlements,
    attributes
  };
};

/**
 @param key feature key
 @param customAttributes user attributes
 @returns if the user is entitled to the given feature and attributes. Attaching the justification if not
 */
const useFeatureEntitlements = (key, customAttributes) => {
  const {
    entitlements,
    attributes
  } = useEntitlementsQueryData(customAttributes);
  const {
    appName
  } = (0, _common.useRootState)();
  return (0, _reduxStore.getFeatureEntitlements)(entitlements, key, attributes, undefined, appName);
};

/**
 @param key permission key
 @param customAttributes user attributes
 @returns if the user is entitled to the given permission and attributes. Attaching the justification if not
 */
exports.useFeatureEntitlements = useFeatureEntitlements;
const usePermissionEntitlements = (key, customAttributes) => {
  const {
    entitlements,
    attributes
  } = useEntitlementsQueryData(customAttributes);
  const {
    appName
  } = (0, _common.useRootState)();
  return (0, _reduxStore.getPermissionEntitlements)(entitlements, key, attributes, undefined, appName);
};

/**
 @param options
 @param customAttributes user attributes
 @returns if the user is entitled to the given feature or permission and attributes (check only one). Attaching the justification if not
 */
exports.usePermissionEntitlements = usePermissionEntitlements;
const useEntitlements = (options, customAttributes) => {
  const {
    entitlements,
    attributes
  } = useEntitlementsQueryData(customAttributes);
  const {
    appName
  } = (0, _common.useRootState)();
  return (0, _reduxStore.getEntitlements)(entitlements, options, attributes, undefined, appName);
};

/**
 @returns an action your can use to detect if the user is entitled to the given feature or permission (check only one).
  Attaching the justification if not
 */
exports.useEntitlements = useEntitlements;
const useEntitlementsActions = () => {
  // this code is duplicated because React is yelling when using useEntitlementsQueryData inside the isEntitledTo function because it's not a hook
  const user = useUserState();
  const entitlements = useEntitlementsState();
  const {
    appName
  } = (0, _common.useRootState)();
  return (0, _react.useMemo)(() => ({
    isEntitledTo: (options, customAttributes) => {
      const attributes = {
        custom: customAttributes,
        jwt: user
      };
      return (0, _reduxStore.getEntitlements)(entitlements, options, attributes, undefined, appName);
    }
  }), [user, entitlements]);
};

/**
 @returns if the option to use entitlements is enabled
 */
exports.useEntitlementsActions = useEntitlementsActions;
const useEntitlementsOptions = () => {
  var _entitlementsOptions$;
  const {
    entitlementsOptions
  } = (0, _common.useShadowDom)();
  const isEntitlementsEnabled = (_entitlementsOptions$ = entitlementsOptions == null ? void 0 : entitlementsOptions.enabled) != null ? _entitlementsOptions$ : false;
  const {
    isEntitledTo
  } = useEntitlementsActions();
  const [verifyIsEntitledFF] = (0, _flags.useFeatureFlags)(['admin-portal-use-is-entitled']);
  return {
    isEntitlementsEnabled,
    isEntitledTo,
    verifyIsEntitledFF
  };
};
exports.useEntitlementsOptions = useEntitlementsOptions;
