// @frontegg/redux-store/node/subscriptions/Plans/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _helpers = require("../../helpers");
var _state = require("./state");
var _default = (store, api, sharedActions) => {
  const setPlansState = state => {
    Object.assign(store.subscriptions.plans, state);
  };
  const resetPlansState = () => {
    (0, _helpers.deepResetState)(store, ['subscriptions', 'plans'], _state.initialState);
  };
  const setPlansError = error => {
    setPlansState({
      loading: false,
      fetching: false,
      error
    });
  };
  const loadPlans = async () => {
    setPlansState({
      loading: true
    });
    try {
      const products = await api.subscriptions.getSubscriptionPlans();
      const plans = products.map(item => {
        var _item$price, _item$price2;
        return {
          id: item.id,
          name: item.name,
          description: item.description,
          price: ((_item$price = item.price) == null ? void 0 : _item$price.amount) || 0,
          currency: ((_item$price2 = item.price) == null ? void 0 : _item$price2.currency) || 'usd',
          recurringInterval: 'month',
          slug: item.slug
        };
      });
      setPlansState({
        fetching: false,
        loading: false,
        plans
      });
    } catch (e) {
      setPlansError((0, _helpers.errorHandler)(e));
    }
  };
  return {
    setPlansState,
    resetPlansState,
    loadPlans
  };
};
exports.default = _default;
