// @frontegg/rest-api/node/subscriptions/providers/stripe/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StripeProviderApi = void 0;
var _constants = require("../../../constants");
var _BaseApiClient = require("../../../BaseApiClient");
class StripeProviderApi extends _BaseApiClient.BaseApiClient {
  constructor(...args) {
    super(...args);
    this.createSubscription = async request => {
      return this.post(_constants.urls.subscriptions.billing.paymentProviders.stripe.subscriptions.v1, request);
    };
    this.getStripeCustomer = async tenantId => {
      return this.get(`${_constants.urls.subscriptions.billing.paymentProviders.stripe.customers.v1}/${tenantId}`);
    };
    this.createStripeCustomer = async request => {
      return this.post(_constants.urls.subscriptions.billing.paymentProviders.stripe.customers.v1, request);
    };
    this.getStripePaymentProviderConfiguration = async () => {
      return this.get(_constants.urls.subscriptions.billing.paymentProviders.stripe.publicConfigurations.v1);
    };
    this.createStripePaymentMethodSetupIntentSecret = async request => {
      return this.post(_constants.urls.subscriptions.billing.paymentProviders.stripe.setupIntents.v1, request);
    };
  }
}
exports.StripeProviderApi = StripeProviderApi;
