// @frontegg/redux-store/node/auth/ForgotPasswordState/actions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _restApi = require("@frontegg/rest-api");
var _interfaces = require("./interfaces");
var _state = require("./state");
var _helpers = require("../../helpers");
const _excluded = ["callback"];
var _default = (store, api, sharedActions) => {
  const actions = sharedActions;
  const setForgotPasswordState = payload => {
    Object.assign(store.auth.forgotPasswordState, payload);
  };
  const resetForgotPasswordState = () => {
    (0, _helpers.deepResetState)(store, ['auth', 'forgotPasswordState'], _state.initialState);
  };
  const forgotPassword = async payload => {
    setForgotPasswordState({
      loading: true
    });
    try {
      var _payload$callback;
      await api.auth.forgotPassword(payload);
      setForgotPasswordState({
        loading: false,
        error: undefined,
        step: _interfaces.ForgotPasswordStep.forgotPasswordEmailSuccess
      });
      (_payload$callback = payload.callback) == null ? void 0 : _payload$callback.call(payload, true);
    } catch (e) {
      var _payload$callback2;
      setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e, 'Unknown error occurred')
      });
      (_payload$callback2 = payload.callback) == null ? void 0 : _payload$callback2.call(payload, false, e);
    }
  };
  const sendPasswordRecoveryEmail = async () => {
    const {
      identifier,
      identifierType
    } = store.auth.forgotPasswordState;
    setForgotPasswordState({
      loading: true
    });
    try {
      await api.auth.resetPasswordViaEmail({
        identifier,
        identifierType
      });
      setForgotPasswordState({
        loading: false,
        error: undefined,
        step: _interfaces.ForgotPasswordStep.forgotPasswordEmailSuccess
      });
    } catch (e) {
      setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e, 'An error occurred while sending recovery email')
      });
    }
  };
  const sendPasswordRecoverySms = async () => {
    const {
      identifier,
      identifierType
    } = store.auth.forgotPasswordState;
    setForgotPasswordState({
      loading: true
    });
    try {
      const response = await api.auth.resetPasswordViaSms({
        identifier,
        identifierType
      });
      setForgotPasswordState({
        loading: false,
        error: undefined,
        sessionId: response.sessionId
      });
    } catch (e) {
      setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e, 'An error occurred while sending recovery sms')
      });
    }
  };
  const verifyPasswordViaSms = async payload => {
    const {
      sessionId
    } = store.auth.forgotPasswordState;
    if (!sessionId) {
      const error = 'Session ID not found in state for verifying password via sms.';
      setForgotPasswordState({
        loading: false,
        error
      });
      return;
    }
    setForgotPasswordState({
      loading: true
    });
    try {
      const response = await api.auth.verifyPasswordViaSms({
        otcToken: payload.otcToken,
        sessionId
      });
      if (!response.userId || !response.token) {
        const error = 'Invalid response from server: missing userId or token';
        throw new Error(error);
      }
      setForgotPasswordState({
        loading: false,
        error: undefined,
        userId: response.userId,
        token: response.token,
        step: _interfaces.ForgotPasswordStep.resetPasswordPage
      });
    } catch (e) {
      setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e, 'An error occurred while verifying password via sms')
      });
    }
  };
  const determinePasswordRecoveryStrategy = async payload => {
    setForgotPasswordState({
      loading: true,
      error: undefined,
      identifier: payload.identifier,
      identifierType: payload.identifierType
    });
    try {
      const strategies = await api.auth.getPasswordRecoveryStrategies();
      const isEmailActive = strategies.some(s => s.strategy === _restApi.PasswordRecoveryStrategyEnum.Email && s.isActive);
      const isSmsActive = strategies.some(s => s.strategy === _restApi.PasswordRecoveryStrategyEnum.Sms && s.isActive);
      if (isEmailActive && isSmsActive) {
        setForgotPasswordState({
          loading: false,
          step: _interfaces.ForgotPasswordStep.passwordRecoverySelector,
          activeStrategies: strategies
        });
      } else if (isEmailActive) {
        await sendPasswordRecoveryEmail();
      } else if (isSmsActive) {
        setForgotPasswordState({
          loading: false,
          step: _interfaces.ForgotPasswordStep.resetPasswordViaSms,
          activeStrategies: strategies
        });
      } else {
        const error = 'No active password recovery methods found.';
        setForgotPasswordState({
          loading: false,
          error,
          activeStrategies: strategies
        });
      }
    } catch (e) {
      setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e, 'An error occurred while determining recovery strategy')
      });
    }
  };
  const resetPassword = async payload => {
    const {
        callback
      } = payload,
      body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded);
    setForgotPasswordState({
      loading: true
    });
    try {
      await api.auth.resetPassword(body);
      setForgotPasswordState({
        loading: false,
        error: undefined,
        step: _interfaces.ForgotPasswordStep.resetPasswordSuccess
      });
      callback == null ? void 0 : callback(true);
    } catch (e) {
      setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e)
      });
      callback == null ? void 0 : callback(false, e);
    }
  };
  const loadPasswordConfig = async payload => {
    setForgotPasswordState({
      loading: true
    });
    try {
      const passwordConfig = await api.auth.loadPasswordConfig(payload);
      setForgotPasswordState({
        loading: false,
        passwordConfig
      });
    } catch (e) {
      setForgotPasswordState({
        loading: false,
        error: (0, _helpers.errorHandler)(e)
      });
    }
  };
  return {
    setForgotPasswordState,
    resetForgotPasswordState,
    forgotPassword,
    resetPassword,
    loadPasswordConfig,
    determinePasswordRecoveryStrategy,
    sendPasswordRecoveryEmail,
    sendPasswordRecoverySms,
    verifyPasswordViaSms
  };
};
exports.default = _default;
